{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cypher Queries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### node list (autophagy reactions as nodes)\n",
    "MATCH(tpa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ev:ReactionLikeEvent)<br>\n",
    "MATCH(ev)<-[:hasEvent]-(pa:Pathway)<br>\n",
    "OPTIONAL MATCH(ev)-[:input]->(pin:PhysicalEntity)<br>\n",
    "OPTIONAL MATCH(ev)-[:output]->(pout:PhysicalEntity)<br>\n",
    "OPTIONAL MATCH(ev)-[:catalystActivity|physicalEntity|regulatedBy|regulator|hasComponent|hasMember|hasCandidate&ast;]->(pot:PhysicalEntity)<br>\n",
    "OPTIONAL MATCH(ev)-[:authored]-(ins:InstanceEdit)-[:author]-(per:Person)<br>\n",
    "RETURN DISTINCT pa.displayName AS PathwayName, ev.displayName AS ReactionName,ev.stId AS ReactomeID,COLLECT(DISTINCT pin.displayName) AS Input,COLLECT(DISTINCT pout.displayName) AS Output,COLLECT(DISTINCT pot.displayName) AS OtherEntities,per.displayName AS Author<br><br>\n",
    "#edge list (edges created from shared PEs)\n",
    "MATCH(pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ro:ReactionLikeEvent)-[:output|catalystActivity|regulator]->(po:PhysicalEntity)\n",
    "MATCH(po)<-[:input|regulatedBy]-(ri:ReactionLikeEvent)<-[:hasEvent&ast;]-(pa)\n",
    "RETURN DISTINCT ro.stId AS OutputReaction,ri.stId AS InputReaction\n",
    "##edge list (edges created from shared PEs by splitting sets into individual components)\n",
    "MATCH(pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output|catalystActivity|regulator]->(po1:PhysicalEntity)\n",
    "MATCH(ro{speciesName:\"Homo sapiens\"})-[:output|catalystActivity|regulator&ast;]->(po2:PhysicalEntity)\n",
    "MATCH(po1)<-[:input|regulatedBy]-(ri1:ReactionLikeEvent)<-[:hasEvent&ast;]-(pa)\n",
    "MATCH(po2)<-[:input|regulatedBy&ast;]-(ri2:ReactionLikeEvent)<-[:hasEvent&ast;]-(pa)\n",
    "RETURN DISTINCT ro.stId AS OutputReaction,\n",
    "//po1.stId AS SharedEntity, po1.schemaClass AS Class,\n",
    "CASE\n",
    "WHEN po1.schemaClass=\"DefinedSet\" \n",
    "THEN ri2.stId\n",
    "ELSE ri1.stId\n",
    "END AS InputReaction\n",
    "###node list (participants as nodes with SimpleEntities eliminated)\n",
    "MATCH(pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input]->(pin:PhysicalEntity)\n",
    "WHERE NOT pin.schemaClass = \"SimpleEntity\"\n",
    "RETURN DISTINCT pin.stId AS Participants\n",
    "UNION\n",
    "MATCH(pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(pout:PhysicalEntity)\n",
    "WHERE NOT pout.schemaClass = \"SimpleEntity\"\n",
    "RETURN DISTINCT pout.stId AS Participants\n",
    "UNION\n",
    "MATCH(pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:catalystActivity|physicalEntity&ast;]->(pcat:PhysicalEntity)\n",
    "WHERE NOT pcat.schemaClass = \"SimpleEntity\"\n",
    "RETURN DISTINCT pcat.stId AS Participants\n",
    "UNION\n",
    "MATCH(pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:regulatedBy|regulator|physicalEntity&ast;]->(preg:PhysicalEntity)\n",
    "WHERE NOT preg.schemaClass = \"SimpleEntity\"\n",
    "RETURN DISTINCT preg.stId AS Participants\n",
    "\n",
    "#### list of events and their following events based on shared entities (simpleEntities & Ub are ignored AND DefinedSets are split)\n",
    "MATCH(pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po1:PhysicalEntity)<br>\n",
    "WHERE NOT (po1.schemaClass=\"SimpleEntity\" OR po1.stId=\"R-HSA-113595\")<br>\n",
    "//WHERE NOT (po1.schemaClass=\"SimpleEntity\" OR po1.stId=\"R-HSA-113595\" OR po1.stId=\"R-HSA-184258\" OR po1.stId=\"R-HSA-3149456\" OR po1.stId=\"R-HSA-917706\")<br>\n",
    "WITH pa, ro, po1<br>\n",
    "MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity&ast;]-(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent&ast;]-(pa)<br>\n",
    "WITH ro, ri1<br>\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction <br>\n",
    "UNION\n",
    "MATCH(pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2:PhysicalEntity{schemaClass:\"DefinedSet\"})-[:output|physicalEntity|hasMember|hasCandidate&ast;]->(po3:PhysicalEntity)<br>\n",
    "WHERE NOT (po2.stId=\"R-HSA-113595\")<br>\n",
    "//WHERE NOT (po1.stId=\"R-HSA-113595\" OR po1.stId=\"R-HSA-184258\" OR po1.stId=\"R-HSA-3149456\" OR po1.stId=\"R-HSA-917706\")<br>\n",
    "WITH pa, ro, po3<br> \n",
    "MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate&ast;]-(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent&ast;]-(pa)<br>\n",
    "WITH ro, ri2<br>\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction<br>\n",
    "\n",
    "#### list of events and their following events based on shared entities (ChemicalDrug & ProteinDrug & simpleEntities & Ub are ignored AND DefinedSets are split)\n",
    "<!-- \n",
    "MATCH(pa:Pathway{stId:\"R-HSA-9679506\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "WITH pa, ro\n",
    "MATCH(ro)-[:output]->(po1:PhysicalEntity)\n",
    "WHERE NOT (po1.schemaClass=\"SimpleEntity\" OR po1.stId=\"R-HSA-113595\")\n",
    "WITH pa, ro, po1\n",
    "MATCH(pa:Pathway{stId:\"R-HSA-9679506\"})-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc2:PhysicalEntity)\n",
    "WITH  pa, ro, po1, ri1, COLLECT(DISTINCT dc2.schemaClass) AS alldc2\n",
    "WHERE NOT (\"ChemicalDrug\" IN (alldc2) OR \"ProteinDrug\" IN (alldc2))\n",
    "WITH ro, po1, ri1\n",
    "MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "WITH ro, ri1\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction\n",
    "ORDER BY ro.stId\n",
    "UNION MATCH(pa:Pathway{stId:\"R-HSA-9679506\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2:PhysicalEntity{schemaClass:\"DefinedSet\"})-[:output|physicalEntity|hasMember|hasCandidate*]->(po3:PhysicalEntity)\n",
    "WHERE NOT (po2.stId=\"R-HSA-113595\")\n",
    "WITH pa, ro, po3\n",
    "MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)\n",
    "WITH ro, ri2\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction\n",
    "ORDER BY ro.stId\n",
    "-->\n",
    "\n",
    "#### list of events and their following events based on shared entities with details of shared entities (ChemicalDrug & ProteinDrug & simpleEntities & Ub are ignored AND DefinedSets are split)\n",
    "<!-- \n",
    "//find reaction connections for all except Set connectors\n",
    "///remove drug-related reactions\n",
    "MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "///check for entity sharing after filtering for simple entity and Ub\n",
    "WITH pa, ro\n",
    "MATCH(ro)-[:output]->(po1:PhysicalEntity)\n",
    "WHERE NOT (po1.schemaClass=\"SimpleEntity\" OR po1.stId=\"R-HSA-113595\" OR po1.schemaClass=\"DefinedSet\" OR po1.schemaClass=\"CandidateSet\")\n",
    "WITH pa, ro, po1\n",
    "MATCH(pa)-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc2:PhysicalEntity)\n",
    "WITH  pa, ro, po1, ri1, COLLECT(DISTINCT dc2.schemaClass) AS alldc2\n",
    "WHERE NOT (\"ChemicalDrug\" IN (alldc2) OR \"ProteinDrug\" IN (alldc2))\n",
    "WITH ro, po1, ri1\n",
    "MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "OPTIONAL MATCH (cacomp:Compartment)<-[:compartment]-(po1)<-[:catalystActivity|physicalEntity*]-(ri1)\n",
    "OPTIONAL MATCH (recomp:Compartment)<-[:compartment]-(po1)<-[:regulator|physicalEntity*]-(reg:Regulation)<-[:regulatedBy|physicalEntity*]-(ri1)\n",
    "WITH ro, ri1, po1, reg, cacomp, recomp\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction, po1.schemaClass AS SharedEntityClass, po1.stId AS SharedEntity\n",
    "ORDER BY ro.stId\n",
    "\n",
    "//find reaction connections for set connectors\n",
    "///remove drug-related reactions\n",
    "UNION MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "///check for entity sharing after filtering for simple entity and Ub\n",
    "WITH pa, ro\n",
    "MATCH(ro)-[:output]->(po2:PhysicalEntity)-[:hasMember|hasCandidate*]->(po3:PhysicalEntity)\n",
    "WHERE NOT (po3.stId=\"R-HSA-113595\") AND (po2.schemaClass=\"DefinedSet\" OR po2.schemaClass=\"CandidateSet\")\n",
    "WITH pa, ro, po3\n",
    "MATCH(pa)-[:hasEvent*]->(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2)\n",
    "OPTIONAL MATCH (cacomp:Compartment)<-[:compartment]-(po3)<-[:catalystActivity|physicalEntity*]-(ri2)\n",
    "OPTIONAL MATCH (recomp:Compartment)<-[:compartment]-(po3)<-[:regulator|physicalEntity*]-(reg:Regulation)<-[:regulatedBy|physicalEntity*]-(ri2)\n",
    "WITH ro, ri2, po3, reg, cacomp, recomp\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction, po3.schemaClass AS SharedEntityClass, po3.stId AS SharedEntity\n",
    "ORDER BY ro.stId\n",
    "-->\n",
    "\n",
    "#### query to get type of input for a reaction - input, output, catalyst and regulator\n",
    "<!-- \n",
    "MATCH(tpa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "WITH ev\n",
    "OPTIONAL MATCH(ev)-[:input]->(pin:PhysicalEntity)\n",
    "OPTIONAL MATCH (pin)-[:hasMember|hasCandidate*]->(pins:PhysicalEntity)\n",
    "WITH ev,pin,pins\n",
    "OPTIONAL MATCH(ev)-[:output]->(pou:PhysicalEntity)\n",
    "OPTIONAL MATCH (pou)-[:hasMember|hasCandidate*]->(pous:PhysicalEntity)\n",
    "WITH ev,pin,pins,pou,pous\n",
    "OPTIONAL MATCH(ev)-[:catalystActivity|physicalEntity*]->(pca:PhysicalEntity)\n",
    "OPTIONAL MATCH(pca)-[:hasMember|hasCandidate*]->(pcas:PhysicalEntity)\n",
    "WITH ev,pin,pins,pou,pous,pca,pcas\n",
    "OPTIONAL MATCH(ev)-[:regulatedBy|physicalEntity*]->(reg:Regulation)-[:regulator]->(pre:PhysicalEntity)\n",
    "OPTIONAL MATCH(pre)-[:hasMember|hasCandidate*]->(pres:PhysicalEntity)\n",
    "WITH ev, reg, \n",
    "COLLECT(pin.stId)+COLLECT(pins.stId) AS i, COLLECT(pou.stId)+COLLECT(pous.stId) AS o, COLLECT(pca.stId)+COLLECT(pcas.stId) AS c, COLLECT(pre.stId)+COLLECT(pres.stId) AS r\n",
    "RETURN DISTINCT ev.stId AS Reaction, i AS Input, o AS Output, c as Catalyst, r AS Regulator, reg.displayName AS Regulation\n",
    "-->\n",
    "\n",
    "\n",
    "#### edge list (edges created from preceding event relation)\n",
    "MATCH(pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<br>\n",
    "MATCH(ev)-[:precedingEvent]->(pe:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<br>\n",
    "RETURN DISTINCT pe.stId AS First_Reaction, ev.stId AS Second_Reaction<br>\n",
    "\n",
    "#### reaction-entity relation list (created for perturbation analysis)\n",
    "MATCH (pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<br>\n",
    "OPTIONAL MATCH (ev)-[:input|catalystActivity|regulatedBy|regulator|physicalEntity]->(PEI:PhysicalEntity)<br>\n",
    "OPTIONAL MATCH (ev)-[:input|catalystActivity|regulatedBy|regulator|hasComponent|hasMember|hasCandidate|repeatedUnit|physicalEntity&ast;]->(ewasI:EntityWithAccessionedSequence)-[:referenceEntity]-(refI:ReferenceGeneProduct)<br>\n",
    "RETURN DISTINCT ev.displayName AS ReactioneName, ev.stId AS ReactionID,\"Input\" AS Category,PEI.schemaClass AS Class,refI.displayName AS EntityName,refI.identifier AS UniprotID<br>\n",
    "UNION MATCH (pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent&ast;]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<br>\n",
    "OPTIONAL MATCH (ev)-[:output|physicalEntity]->(PEO:PhysicalEntity)<br>\n",
    "OPTIONAL MATCH (ev)-[:output|hasComponent|hasMember|hasCandidate|repeatedUnit|physicalEntity&ast;]->(ewasO:EntityWithAccessionedSequence)-[:referenceEntity]-(refO:ReferenceGeneProduct)<br>\n",
    "RETURN DISTINCT ev.displayName AS ReactioneName, ev.stId AS ReactionID,\"Ouput\" AS Category,PEO.schemaClass AS Class,refO.displayName AS EntityName,refO.identifier AS UniprotID<br>\n",
    "\n",
    "#### METABOLISM: list of events and their following events based on shared entities (specific simpleEntities & Ub are ignored AND DefinedSets are split)\n",
    "MATCH(pa:Pathway{stId:\"R-HSA-1430728\"})-[:hasEvent&ast;]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<br>\n",
    "MATCH(ro)-[:input|output|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]->(po1:PhysicalEntity)-[:referenceEntity]->(ref0:ReferenceMolecule)<br>\n",
    "WHERE NOT ANY(x IN ref0.identifier WHERE x IN [\"15377\",\"16761\",\"15422\",\"16027\",\"15996\",\"17552\",\"17659\",\"15713\",\"15378\",\"16234\",\"18367\",\"29888\",\"13392\",\"18009\",\"16908\",\"16238\",\"17877\",\"15346\",\"15379\",\"16526\",\"15846\",\"17345\",\"16174\",\"17713\",\"16284\",\"28862\",\"16192\",\"16497\",\"17625\",\"15713\",\"16695\",\"18009\"])<br>\n",
    "WITH pa, ro, po1<br>\n",
    "MATCH(po1)<-[:output]-(ro)<br>\n",
    "WHERE NOT (po1.stId=\"R-HSA-113595\")<br>\n",
    "WITH pa, ro, po1<br>\n",
    "MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent&ast;]-(pa)<br>\n",
    "WITH ro, ri1<br>\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction<br>\n",
    "UNION MATCH(pa:Pathway{stId:\"R-HSA-1430728\"})-[:hasEvent&ast;]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2:PhysicalEntity{schemaClass:\"DefinedSet\"})-[:output|physicalEntity|hasMember|hasCandidate*]->(po3:PhysicalEntity)<br>\n",
    "WHERE NOT (po2.stId=\"R-HSA-113595\")<br>\n",
    "WITH pa, ro, po3<br>\n",
    "MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent&ast;]-(pa)<br>\n",
    "WITH ro, ri2<br>\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction<br>\n",
    "\n",
    "#### Disease-Reaction map\n",
    "MATCH (pa:TopLevelPathway{stId:\"R-HSA-1643685\"})-[:hasEvent&ast;]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:disease]-(di:Disease)<br>\n",
    "RETURN DISTINCT ro.stId AS ReactionID, ro.displayName AS ReactionName, di.identifier AS DiseaseID, di.displayName AS DiseaseName<br>\n",
    "UNION MATCH (pa:TopLevelPathway{stId:\"R-HSA-1643685\"})-[:hasEvent&ast;]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:disease]-(di:Disease)<br>\n",
    "MATCH (ro)-[:normalReaction]->(nro:ReactionLikeEvent)<br>\n",
    "RETURN DISTINCT nro.stId AS ReactionID, nro.displayName AS ReactionName, di.identifier AS DiseaseID, di.displayName AS DiseaseName<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Network analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import networkx as nx\n",
    "#%matplotlibe notebook\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sb\n",
    "import random\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def graphBuilder(nodes,edges):\n",
    "    G = nx.MultiDiGraph() #create a new graph object\n",
    "    G.add_nodes_from(nodes) #add nodes\n",
    "    G.add_edges_from(edges) #add edges\n",
    "    G.name = ''\n",
    "    print(nx.info(G))\n",
    "    return(G)\n",
    "def nodeNumberer(G):\n",
    "    pos = nx.spring_layout(G)\n",
    "    nodeLabelDict = {} #create numbered labels for better visualization\n",
    "    nodeLabelVal = 0\n",
    "    for key in pos.keys():\n",
    "        nodeLabelVal+=1\n",
    "        nodeLabelDict[key]=nodeLabelVal\n",
    "        x,y=pos[key]\n",
    "#        plt.text(x,y,s=str(nodeLabelVal))\n",
    "    return nodeLabelDict\n",
    "def pathExplorer(G,source,target,c=None):\n",
    "    pos = nx.spring_layout(G)\n",
    "#    nx.draw(G,pos,node_color='y')\n",
    "#    nx.draw_networkx_nodes(G,pos,nodelist=[source,target],node_color='r')\n",
    "    nodeLabelDict = {} #create numbered labels for better visualization\n",
    "    nodeLabelVal = 0\n",
    "    for key in pos.keys():\n",
    "        nodeLabelVal+=1\n",
    "        nodeLabelDict[key]=nodeLabelVal\n",
    "        x,y=pos[key]\n",
    "#        plt.text(x,y,s=str(nodeLabelVal))\n",
    "    #nodeLabelDict=nodeNumberer(G)\n",
    "#    legLab = [str(nodeLabelDict[key])+'->'+str(key) for key in nodeLabelDict]\n",
    "#    plt.legend(legLab,loc='best',bbox_to_anchor=(3,1),ncol=5) #parameter to change position of labels\n",
    "    pathCount = 0 #trace path between source and target\n",
    "    pathTracer = []\n",
    "    for path in nx.all_simple_paths(G,source=source,target=target,cutoff=c):\n",
    "        #print(path)\n",
    "        #print(list(zip(path,path[1:])))\n",
    "        pathEdges = list(zip(path,path[1:]))\n",
    "#        nx.draw_networkx_nodes(G,pos,nodelist=path,node_color='r')\n",
    "#        nx.draw_networkx_edges(G,pos,edgelist=pathEdges,edge_color='r',width=10)\n",
    "        pathCount = pathCount+1\n",
    "        Path_Map=str(nodeLabelDict[path[0]])\n",
    "        for n in list(path[1:]):\n",
    "            Path_Map=Path_Map+'-->'+str(nodeLabelDict[n])\n",
    "        pathTracer.append(Path_Map)\n",
    "    print('Number of paths from '+source+'('+str(nodeLabelDict[source])+')'+\n",
    "          ' to '+target+'('+str(nodeLabelDict[target])+')'+' is '+str(len(pathTracer)))\n",
    "#    print(pathTracer)\n",
    "    plt.show()\n",
    "def adjacencyMatrixViewer(G,nodes):\n",
    "    #create adjacency matrix\n",
    "    sortedNodes = sorted([node for node,degree in G.degree()],reverse=True) #sorted based on descending degrees (degree=number of edges from the node)\n",
    "    adjMat = nx.adjacency_matrix(G,nodelist=sortedNodes)\n",
    "    edgeMap = adjMat.todense()\n",
    "    #plot heatmap\n",
    "    fig, ax = plt.subplots()\n",
    "    im = ax.imshow(edgeMap) #imshow is the heatmap plot\n",
    "    ax.set_title('Adjacency Matrix (descending degrees)')\n",
    "    ax.set_xlabel('Following Reaction Number')\n",
    "    ax.set_ylabel('Reaction Number')\n",
    "    ax.set_ylim(ax.get_ylim()[::-1])\n",
    "    fig.tight_layout()\n",
    "    plt.show()\n",
    "def edgeListTrimmer(KO_Reactions,G):\n",
    "    ##function that eliminates a list of KO reactions and returns the remaining edges\n",
    "    ##KO_Reactions is a list of list of tuples\n",
    "    All_Edges=list(G.edges())\n",
    "    All_KO_Reactions=[j for i in KO_Reactions for j in list(set(i))] #grouping all KO reactions\n",
    "    #deriving all the edges resulting from KO reactions \n",
    "    All_KO_Edges=[]\n",
    "    for rxn in All_KO_Reactions:\n",
    "        KO_Edges=list(nx.dfs_edges(G,source=rxn))\n",
    "        for i in KO_Edges:\n",
    "            All_KO_Edges.append(i)\n",
    "    #printing deleted edges as numbers\n",
    "##commented to avoid pos variable from nodeNumberer\n",
    "##uncomment to display deleted edges in KO analysis\n",
    "#    nodeLabelDict=nodeNumberer(G)\n",
    "#    Deleted_Edges=[]\n",
    "#    for i in All_KO_Edges:\n",
    "#        if i not in Deleted_Edges:\n",
    "#            Deleted_Edges.append(i)\n",
    "##print deleted edges\n",
    "#    print('Deleted Edges: ')\n",
    "#    for i in Deleted_Edges:\n",
    "#        print((str(nodeLabelDict[i[0]])+'-->'+str(nodeLabelDict[i[1]])+'\\t\\t'),(str(i[0])+'-->'+str(i[1])))\n",
    "    New_Edge_List=[edge for edge in All_Edges if edge not in All_KO_Edges]\n",
    "    return All_KO_Edges,New_Edge_List\n",
    "def heatmaPlotter(data,x_Labels,y_Labels,Title,x_Axis_Name,y_Axis_Name):\n",
    "    data=np.asarray(data)\n",
    "    #Total_data=np.sum(data)\n",
    "    #plot generator\n",
    "    fig, ax = plt.subplots()\n",
    "    plt.rcParams[\"figure.figsize\"] = (30,30)\n",
    "    im = ax.imshow(data,cmap='YlOrRd')\n",
    "    #show all ticks\n",
    "    ax.set_xticks(np.arange(len(x_Labels)))\n",
    "    ax.set_yticks(np.arange(len(y_Labels)))\n",
    "    #label them with the respective list entries\n",
    "    ax.set_xticklabels(x_Labels)\n",
    "    ax.set_yticklabels(y_Labels)\n",
    "    #rotate the tick labels and set their alignment.\n",
    "    plt.setp(ax.get_xticklabels(), rotation=45, ha=\"right\",\n",
    "             rotation_mode=\"anchor\")\n",
    "#    #Loop over data dimensions and create text annotations.\n",
    "#    for i in range(len(y_Labels)):\n",
    "#        for j in range(len(x_Labels)):\n",
    "#            text = ax.text(j, i, data[i, j],\n",
    "#                           ha=\"center\", va=\"center\", color=\"k\")\n",
    "    #name labels and title\n",
    "    ax.set_title(Title)\n",
    "    ax.set_xlabel(x_Axis_Name)\n",
    "    ax.set_ylabel(y_Axis_Name)\n",
    "#    fig.tight_layout()\n",
    "    plt.show()\n",
    "    return fig\n",
    "def pathCountHeatmap(G,Source_List,Target_List,Title):\n",
    "#counts number of paths from source to target\n",
    "    Paths_Matrix=[]\n",
    "    for i in Source_List:\n",
    "        Path_Count=[]\n",
    "        for j in Target_List:\n",
    "            All_Paths=nx.all_simple_paths(G,source=i,target=j)\n",
    "            Path_Count.append(len([path for path in All_Paths]))\n",
    "        Paths_Matrix.append(Path_Count)\n",
    "    Source_Labels=[Path_Dict[i][0] for i in Source_List]\n",
    "    Target_Labels=[Path_Dict[i][0] for i in Target_List]\n",
    "    heatmaPlotter(Paths_Matrix,Target_Labels,Source_Labels,Title,'Target Node','Source Node')\n",
    "    return Total_Paths\n",
    "def geneKOHeatmap(UniprotIDs,Title):\n",
    "#funcion to plot heatmap of perturbed network\n",
    "    KO_Genes=UniprotIDs\n",
    "    KO_Reactions=[Protein_Dict[protein] for protein in KO_Genes]\n",
    "    All_KO_Edges,New_Edge_List=edgeListTrimmer(KO_Reactions,G)\n",
    "#    All_Dnstrm_Nodes=KO_Reactions[0]+[j for i in New_Edge_List for j in i]\n",
    "#    All_KO_Nodes=[i for i in list(set(All_Dnstrm_Nodes)) if All_Dnstrm_Nodes.count(i)==G.degree[i]]\n",
    "#    KO_Node_IDs=[i for i in nodeIDs if i not in All_KO_Nodes]\n",
    "    G_Pert = nx.MultiDiGraph()\n",
    "#    G_Pert.add_nodes_from(KO_Node_IDs)\n",
    "    G_Pert.add_nodes_from(nodeIDs)\n",
    "    G_Pert.add_edges_from(New_Edge_List)\n",
    "    Total_Paths=pathCountHeatmap(G_Pert,Source_List,Target_List,Title)\n",
    "    return Total_Paths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Source and target nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Autophagy\n",
    "MA_SourceTarget=['R-HSA-5672817','R-HSA-5682388']\n",
    "CMA_SourceTarget=['R-HSA-9615721','R-HSA-9626060']\n",
    "MI_SourceTarget=['R-HSA-9615721','R-HSA-9631065']\n",
    "ParkinMit_SourceTarget=['R-HSA-5205661','R-HSA-5205663']\n",
    "ReceptorMit_SourceTarget=['R-HSA-8948039','R-HSA-8959571']\n",
    "Aggrephagy_SourceTarget=['R-HSA-9639698','R-HSA-9646685']\n",
    "Lipophagy_SourceTarget=['R-HSA-9613352','R-HSA-9613666']\n",
    "Pexophagy_SourceTarget=['R-HSA-9664850','R-HSA-9664880']\n",
    "#Immune system\n",
    "ImmuneSystem_Sources=['R-HSA-975866','R-HSA-1678923','R-HSA-1236958','R-HSA-1236939','R-HSA-2130627']\n",
    "ImmuneSystem_Targets=['R-HSA-975852','R-HSA-168166','R-HSA-1168633','R-HSA-1168636','R-HSA-2025882','R-HSA-2730894']\n",
    "#Signaling of RTK\n",
    "RTKSignaling_SourceTarget=['R-HSA-177946','R-HSA-177923']\n",
    "#Metabolism\n",
    "Metabolism_SourceTarget=['R-HSA-70420','R-HSA-70369']\n",
    "#SARS\n",
    "SARS_SourceTarget=['R-HSA-9678128','R-HSA-9685542']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Graph building and path analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Autophagy:manually added links that cannot be logically created\n",
    "#hubPreced=['R-HSA-5205663','R-HSA-8959571','R-HSA-9646685','R-HSA-9613666','R-HSA-9664880','R-HSA-5682385']\n",
    "#hubEvent='R-HSA-5682388'\n",
    "#manuallyAddedLinks=[[rxn,hubEvent] for rxn in hubPreced]\n",
    "\n",
    "#set source and target\n",
    "source = CMA_SourceTarget[0]\n",
    "target = CMA_SourceTarget[1]\n",
    "#source = ImmuneSystem_Sources[1]\n",
    "#target = ImmuneSystem_Targets[1]\n",
    "\n",
    "#read file - reaction nodes & edges created from preceding event relation\n",
    "PathwayName='Autophagy'\n",
    "\n",
    "#path = \"\\Users\\tvarusai\\Documents\\Win2Mac_Full\\EBI_Work\\Reactome\\Network Analysis Project\\\"\n",
    "\n",
    "notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "path=notebook_path.rsplit('/',1)\n",
    "path=path[0]+'/'\n",
    "\n",
    "nodeList = PathwayName+'_ReactionsAsNodes.csv'\n",
    "edgeList = PathwayName+'_ReactionsAsEdges_PrecedEvent.csv'\n",
    "with open(path+nodeList,'r') as nodeCSV: #open the file\n",
    "    nodeReader = csv.reader(nodeCSV) #read the csv\n",
    "    nodes = [x for x in nodeReader][1:] #retrieve the data as a list and remove header row\n",
    "nodeIDs = [x[2] for x in nodes] #get only the reaction IDs\n",
    "with open(path+edgeList,'r') as edgeCSV:\n",
    "    edgeReader = csv.reader(edgeCSV)\n",
    "    edges = [x for x in edgeReader][1:]\n",
    "    edgesTup = tuple(edges)\n",
    "edgesPreced = edgesTup\n",
    "print('Edges created from preceding event relation')\n",
    "plt.figure(figsize=(5,5))\n",
    "G = graphBuilder(nodeIDs,edgesPreced)\n",
    "pathExplorer(G,source,target,c=25)\n",
    "adjacencyMatrixViewer(G,nodeIDs)\n",
    "\n",
    "#read file -  edges created from shared PEs\n",
    "#edgeList = PathwayName+'_ReactionsAsEdges_SharedIO_SplitSets_SimpleEnityFalse_UbFalse.csv'\n",
    "edgeList = 'b'+PathwayName+'_ReactionsAsEdges_SharedIO_SplitSets_SimpleEnityFalse_UbFalse.csv'\n",
    "with open(path+edgeList,'r') as edgeCSV:\n",
    "    edgeReader = csv.reader(edgeCSV)\n",
    "    edges = [x for x in edgeReader][1:]\n",
    "    edgesTup = tuple([[i[0],i[1]] for i in edges])\n",
    "    edgeSharedType = [i[2] for i in edges]\n",
    "    #edgesTup = tuple(edges)\n",
    "edgesSharedPE = edgesTup\n",
    "##manual addition\n",
    "#edgesSharedPE = edgesSharedPE + tuple(manuallyAddedLinks)  #combining links generated logically with manual ones\n",
    "\n",
    "print('Edges created from shared PEs')\n",
    "plt.figure(figsize=(5,5))\n",
    "G = graphBuilder(nodeIDs,edgesSharedPE)\n",
    "pathExplorer(G,source,target,c=25)\n",
    "adjacencyMatrixViewer(G,nodeIDs)\n",
    "\n",
    "#read file - edges created from preceding event relation and shared PEs\n",
    "print('Edges created from preceding event relation and shared PEs by splitting sets')\n",
    "edgeDiff = [i for i in edgesPreced+edgesSharedPE if i not in edgesPreced]\n",
    "allEdges = edgesPreced + tuple(edgeDiff)\n",
    "\n",
    "plt.figure(figsize=(5,5))\n",
    "G = graphBuilder(nodeIDs,allEdges)\n",
    "pathExplorer(G,source,target,c=25)\n",
    "adjacencyMatrixViewer(G,nodeIDs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set(edgeSharedType)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edgeDiff\n",
    "AuthorsDict = {x[2]:x[6] for x in nodes}\n",
    "[[i[0],i[1],list(set([AuthorsDict[j] for j in [i[0],i[1]]]))] for i in edgeDiff]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "b='=HYPERLINK(\"https://reactome.org/content/detail/' #base url for Reactome website\n",
    "m='\",\"'\n",
    "e='\")'\n",
    "AuthorsDict={x[2]:x[6] for x in nodes}\n",
    "WriteData=[[b+str(i[0])+m+i[0]+e,b+str(i[1])+m+i[1]+e,list(set([AuthorsDict[j] for j in [i[0],i[1]]]))] for i in edgeDiff]\n",
    "\n",
    "OutFilPat = 'C:\\\\Users\\\\tvarusai\\\\EBI_Work\\\\Reactome\\\\Network Analysis Project\\\\Reaction_Connect_Project\\\\'\n",
    "FilNam = PathwayName+'_Missed_Connections.csv'\n",
    "with open(OutFilPat+FilNam,mode='w',newline='') as fil:\n",
    "    writ=csv.writer(fil)\n",
    "    writ.writerow(['First Reaction','Second Reaction','Author(s)'])\n",
    "    writ.writerows(WriteData)\n",
    "    #[writ.writerow([i[0],i[1]]) for i in edgeDiff]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#identifying protein signaling hubs\n",
    "G_Degrees=list(G.degree())\n",
    "Max_Degree=G_Degrees[max(range(0,len(G_Degrees)-1),key=lambda r: G_Degrees[r][1])]\n",
    "Min_Degree=G_Degrees[min(range(0,len(G_Degrees)-1),key=lambda r: G_Degrees[r][1])]\n",
    "print('The node with highest edges is: ',Max_Degree[0])\n",
    "print('The number of edges in this node is: ',Max_Degree[1])\n",
    "print('The node with lowest edges is: ',Min_Degree[0])\n",
    "print('The number of edges in this node is: ',Min_Degree[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create pathway-reaction-protein mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import reaction-pathway list\n",
    "RPMap_File = PathwayName+'_ReactionsAsNodes.csv'\n",
    "with open(path+RPMap_File,'r') as Map_CSV: #open the file\n",
    "    Map_Reader = csv.reader(Map_CSV) #read the csv\n",
    "    RPMapping = [x for x in Map_Reader][1:] #retrieve the data as a list and remove header row\n",
    "#import reaction-protein relation list\n",
    "Map_File=PathwayName+'_ReactionEntityRelations.csv'\n",
    "with open(path+Map_File,'r') as Map_CSV:\n",
    "    Map_Reader = csv.reader(Map_CSV)\n",
    "    Mapping = [x for x in Map_Reader][1:]\n",
    "Mapping[0]\n",
    "\n",
    "#reaction->pathway dictionary\n",
    "Path_Dict={}\n",
    "for rel in RPMapping:\n",
    "    if rel[0] in Path_Dict:\n",
    "        Path_Dict[rel[2]].append(rel[0])\n",
    "    else:\n",
    "        Path_Dict[rel[2]]=[rel[0]]\n",
    "print('reaction->pathway dictionary length: '+str(len(Path_Dict)))\n",
    "#reaction->protein dictionary\n",
    "React_Dict={}\n",
    "for rel in Mapping:\n",
    "    if rel[1] in React_Dict:\n",
    "        React_Dict[rel[1]].append(rel[5])\n",
    "    else:\n",
    "        React_Dict[rel[1]]=[rel[5]]\n",
    "print('reaction->protein dictionary length: '+str(len(React_Dict)))\n",
    "#protein->reaction based dictionary\n",
    "Protein_Dict={}\n",
    "for rel in Mapping:\n",
    "    if rel[5] in Protein_Dict:\n",
    "        Protein_Dict[rel[5]].append(rel[1])\n",
    "    else:\n",
    "        Protein_Dict[rel[5]]=[rel[1]]\n",
    "print('protein->reaction dictionary length: '+str(len(Protein_Dict)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Perturbation analysis: KO effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "KO_Genes=['O75385','Q9C0C7'] #ULK1,AMBRA1\n",
    "#KO_Genes=['Q9C0C7'] #AMBRA1\n",
    "KO_Genes=['O75385'] #ULK1\n",
    "#KO_Genes=['Q14204'] #Dynein motor\n",
    "#KO_Genes=['Q13501'] #SQSTM1\n",
    "#KO_Genes=['O14672'] #ADAM10\n",
    "#KO_Genes=['P35557']\n",
    "#KO_Genes=['Q9NWZ3','O14920','Q14765'] #IRAK4,IKBKB,STAT4\n",
    "KO_Genes=['Q14765'] #STAT4\n",
    "KO_Genes=['P07711'] #CTSL\n",
    "KO_Reactions=[Protein_Dict[protein] for protein in KO_Genes]\n",
    "nodeLabelDict=nodeNumberer(G)\n",
    "KO_Reactions_List=[nodeLabelDict[r] for q in KO_Reactions for r in q]\n",
    "print('KO proteins are: ',KO_Genes)\n",
    "All_KO_Edges,New_Edge_List=edgeListTrimmer(KO_Reactions,G)\n",
    "KO_Reactions_Group=KO_Reactions+All_KO_Edges #grouping directly involved reactions and indirectly involved edges\n",
    "All_KO_Reactions=set([i for j in KO_Reactions_Group for i in j])\n",
    "\n",
    "#Removing nodes that are isolated\n",
    "#All_Dnstrm_Nodes=KO_Reactions[0]+[j for i in New_Edge_List for j in i] #list of all the affected nodes (directly and indirectly)\n",
    "#All_KO_Nodes=[i for i in list(set(All_Dnstrm_Nodes)) if All_Dnstrm_Nodes.count(i)==G.degree[i]] #list of all isolated node that need to be removed\n",
    "#KO_Node_IDs=[i for i in nodeIDs if i not in All_KO_Nodes] #created new node list after removing KO nodes\n",
    "\n",
    "print('Intact graph')\n",
    "G = graphBuilder(nodeIDs,allEdges)\n",
    "pathExplorer(G,source,target,c=25)\n",
    "\n",
    "print('Perturbed graph')\n",
    "print('Deleted nodes: ',sorted([nodeLabelDict[i] for i in All_KO_Reactions]))\n",
    "#G_Pert = graphBuilder(KO_Node_IDs,New_Edge_List)\n",
    "G_Pert=graphBuilder(nodeIDs,New_Edge_List)\n",
    "pathExplorer(G_Pert,source,target,c=25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#creating graph of paths\n",
    "Path_List=[]\n",
    "Path_Edge_List=[]\n",
    "for i in nx.all_simple_paths(G,source=source,target=target,cutoff=25):\n",
    "        n=list(zip(i,i[1:]))\n",
    "        Path_Edge_List.append(n)\n",
    "        Path_List.append(i)\n",
    "\n",
    "\n",
    "Path_Nodes=[nodeLabelDict[j] for i in Path_List for j in i]\n",
    "Path_Edges=[(nodeLabelDict[j[0]],nodeLabelDict[j[1]]) for i in Path_Edge_List for j in i]\n",
    "\n",
    "All_KO_Reaction_Nums=[nodeLabelDict[i] for i in list(All_KO_Reactions)]\n",
    "\n",
    "Del_Nodes=[]\n",
    "for i in Path_Nodes:\n",
    "    if i in All_KO_Reaction_Nums:\n",
    "        Del_Nodes.append(i)\n",
    "Del_Nodes=set(Del_Nodes)    \n",
    "\n",
    "\n",
    "#disease map\n",
    "#import disease-reaction map\n",
    "DRMap_File = 'DiseaseReactionMap.csv'\n",
    "with open(path+DRMap_File,'r') as Map_CSV: #open the file\n",
    "    Map_Reader = csv.reader(Map_CSV) #read the csv\n",
    "    DRMapping = [x for x in Map_Reader][1:] #retrieve the data as a list and remove header row\n",
    "    \n",
    "#disease->reaction dictionary\n",
    "ReactDis_Dict={}\n",
    "for rel in DRMapping:\n",
    "    if rel[0] in ReactDis_Dict:\n",
    "        ReactDis_Dict[rel[0]].append(rel[3])\n",
    "    else:\n",
    "        ReactDis_Dict[rel[0]]=[rel[3]]\n",
    "    \n",
    "All_Dis_Nodes=[nodeLabelDict[i] for i in ReactDis_Dict if i in nodeLabelDict]\n",
    "Dis_Nodes=[]\n",
    "for i in Path_Nodes:\n",
    "    if i in All_Dis_Nodes:\n",
    "        Dis_Nodes.append(i)\n",
    "Dis_Nodes=set(Dis_Nodes) \n",
    "\n",
    "PathG=nx.MultiDiGraph()\n",
    "PathG.add_nodes_from(Path_Nodes)\n",
    "PathG.add_edges_from(Path_Edges)\n",
    "#print(nx.info(PathG))    \n",
    "pos = nx.kamada_kawai_layout(PathG)\n",
    "plt.figure(figsize=(10,10))\n",
    "nx.draw(PathG,pos,node_color='b',with_labels=True,font_color='w')\n",
    "nx.draw_networkx_nodes(PathG,pos,nodelist=Del_Nodes,node_color='r')\n",
    "#nx.draw_networkx_nodes(PathG,pos,nodelist=Dis_Nodes,node_color='k')\n",
    "nx.draw_networkx_nodes(PathG,pos,nodelist=set([136,102]),node_color='y')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#KO_Genes=[['O75385'],['O75385','Q9C0C7'],['O75385','Q9C0C7','P42345'],['O75385','Q9C0C7','P42345','Q14457']] #ULK1,ULK1+AMBR1,ULK1+AMBR1+mTOR,ULK1+AMBR1+mTOR+BCN1\n",
    "#KO_Genes=[['Q14457'],['Q14457','P13473'],['Q14457','P13473','P68104'],['Q14457','P13473','P68104','O60260']]\n",
    "#KO_Genes=[['P00533'],['P42336']] #EGFR #PI3K\n",
    "##all proteins\n",
    "#KO_Genes=[[i] for i in list(Protein_Dict.keys())]\n",
    "##random proteins\n",
    "num=3\n",
    "KO_Genes=[]\n",
    "n=[]\n",
    "RandGenes=random.choices(list(Protein_Dict.keys()),k=num)\n",
    "for i in RandGenes:\n",
    "    n=n+[i]\n",
    "    KO_Genes.append(n)\n",
    "    #KO_Genes.append([i])\n",
    "\n",
    "KO_Genes=['Q9NWZ3','O14920'] #IRAK4,IKBKB\n",
    "    \n",
    "Pathway_Freq_List=[]\n",
    "for genes in KO_Genes:\n",
    "    #KO_Reactions=[Protein_Dict[protein] for protein in genes]\n",
    "    KO_Reactions=[Protein_Dict[genes]]\n",
    "    All_KO_Edges,New_Edge_List=edgeListTrimmer(KO_Reactions,G)\n",
    "    KO_Reactions_Group=KO_Reactions+All_KO_Edges #grouping directly involved reactions and indirectly involved edges\n",
    "    All_KO_Reactions=set([i for j in KO_Reactions_Group for i in j])\n",
    "    Pathway_KO_List=[Path_Dict[i][0] for i in All_KO_Reactions]\n",
    "    Pathway_KO_Dict={}\n",
    "    for i in Pathway_KO_List:\n",
    "        if i in Pathway_KO_Dict:\n",
    "            Pathway_KO_Dict[i]+=1\n",
    "        else:\n",
    "            Pathway_KO_Dict[i]=1\n",
    "    Pathway_Freq=[]\n",
    "    All_Pathways=[i[0] for i in list(Path_Dict.values())]\n",
    "    for pathway in set(All_Pathways):\n",
    "        if pathway in Pathway_KO_Dict:\n",
    "            count=Pathway_KO_Dict[pathway]\n",
    "        else:\n",
    "            count=0\n",
    "        Pathway_Freq.append(count)\n",
    "    Pathway_Freq_List.append(Pathway_Freq)\n",
    "#transposing the data\n",
    "Pathway_Freq_List=np.array(Pathway_Freq_List).T.tolist()\n",
    "\n",
    "\n",
    "x_Labels=[i for i in KO_Genes]\n",
    "y_Labels=[i for i in set(All_Pathways)]\n",
    "\n",
    "#fig=heatmaPlotter(Pathway_Freq_List,x_Labels,y_Labels,PathwayName+' network perturbation','KO genes','# affected reactions')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_Labels=[i for i in KO_Genes]\n",
    "KO_Heatmap=sb.heatmap(Pathway_Freq_List,cmap='YlOrRd')\n",
    "#sb.set(font_scale=0.5)\n",
    "plt.xlabel('KO genes')\n",
    "plt.ylabel('# affected reactions')\n",
    "plt.figure(figsize=(5,20))\n",
    "KO_Heatmap.set(yticklabels=y_Labels)\n",
    "KO_Heatmap.set(xticklabels=x_Labels)\n",
    "KO_Heatmap.set_xticklabels(KO_Heatmap.get_xticklabels(), rotation=90)\n",
    "KO_Heatmap.set_yticklabels(KO_Heatmap.get_yticklabels(), rotation=0)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#check if perturbation has any disease effects downstream\n",
    "#import disease-reaction map\n",
    "DRMap_File = 'DiseaseReactionMap.csv'\n",
    "with open(path+DRMap_File,'r') as Map_CSV: #open the file\n",
    "    Map_Reader = csv.reader(Map_CSV) #read the csv\n",
    "    DRMapping = [x for x in Map_Reader][1:] #retrieve the data as a list and remove header row\n",
    "\n",
    "#all reaction->disease dictionary\n",
    "ReactDis_Dict={}\n",
    "for rel in DRMapping:\n",
    "    if rel[0] in ReactDis_Dict:\n",
    "        ReactDis_Dict[rel[0]].append(rel[3])\n",
    "    else:\n",
    "        ReactDis_Dict[rel[0]]=[rel[3]]\n",
    "        \n",
    "#all disease->reaction dictionary\n",
    "DisReact_Dict={}\n",
    "for i in DRMapping:\n",
    "    if i[3] in DisReact_Dict:\n",
    "        DisReact_Dict[i[3]].append(i[0])\n",
    "    else:\n",
    "        DisReact_Dict[i[3]]=[i[0]]\n",
    "        \n",
    "#disease->protein dictionary\n",
    "ProtDis_Dict={}\n",
    "for i in ReactDis_Dict:\n",
    "    if i in React_Dict:\n",
    "        for j in React_Dict[i]:\n",
    "            ProtDis_Dict[j]=ReactDis_Dict[i]\n",
    "        \n",
    "DisList=[]\n",
    "for i in All_KO_Reactions:\n",
    "#for i in React_Dict:\n",
    "    if i in ReactDis_Dict:\n",
    "        for j in ReactDis_Dict[i]:\n",
    "            DisList.append(j)\n",
    "\n",
    "print('Diseases involved: ',set(DisList))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##create disease pathway heatmap\n",
    "#import pathway map file\n",
    "#import disease-reaction map\n",
    "PathwayMap_File = PathwayName+'_PathwayMap.csv'\n",
    "with open(path+PathwayMap_File,'r') as Map_CSV: #open the file\n",
    "    Map_Reader = csv.reader(Map_CSV) #read the csv\n",
    "    PathwayMapping = [x for x in Map_Reader][1:] #retrieve the data as a list and remove header row\n",
    "#create pathway heirarchy dict\n",
    "PathwayH_Dict={}\n",
    "for i in PathwayMapping:\n",
    "    if i[2] in PathwayH_Dict:\n",
    "        PathwayH_Dict[i[2]].append(i[4])\n",
    "    else:\n",
    "        PathwayH_Dict[i[2]]=[i[4]]\n",
    "\n",
    "\n",
    "DisPat_List=[]\n",
    "for i in sorted(list(DisReact_Dict.keys())):\n",
    "    y=[]\n",
    "    for j in set(list(PathwayH_Dict.keys())):\n",
    "        w=0\n",
    "        for k in PathwayH_Dict[j]:\n",
    "            if k in [v for v in DisReact_Dict[i]]:\n",
    "                w+=1\n",
    "        y.append(w)\n",
    "    DisPat_List.append(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_Labels=[i for i in sorted(list(DisReact_Dict.keys()))]\n",
    "y_Labels=[i for i in set(list(PathwayH_Dict.keys()))]\n",
    "plt.figure(figsize=(20,20))\n",
    "#DP_Heatmap=sb.heatmap(np.array(DisPat_List).T.tolist(),cmap='YlOrRd')\n",
    "DP_Heatmap=sb.heatmap(DisPat_List,cmap='YlOrRd')\n",
    "sb.set(font_scale=1)\n",
    "plt.xlabel('Diseases')\n",
    "plt.ylabel('Pathways')\n",
    "DP_Heatmap.set(yticklabels=y_Labels)\n",
    "DP_Heatmap.set(xticklabels=x_Labels)\n",
    "DP_Heatmap.set_xticklabels(DP_Heatmap.get_xticklabels(), rotation=90)\n",
    "DP_Heatmap.set_yticklabels(DP_Heatmap.get_yticklabels(), rotation=0)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig=KO_Heatmap.get_figure()\n",
    "#fig.savefig(path+PathwayName+'_All_Pert.png',bbox_inches = 'tight')\n",
    "fig.savefig(path+'Autophagy_ImmuneSystem_MultiPert.png',bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pathway function-based analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Source_List=[MA_SourceTarget[0],CMA_SourceTarget[0],MI_SourceTarget[0],ParkinMit_SourceTarget[0],\n",
    "             ReceptorMit_SourceTarget[0],Aggrephagy_SourceTarget[0],Lipophagy_SourceTarget[0],\n",
    "            Pexophagy_SourceTarget[0]]\n",
    "Target_List=[MA_SourceTarget[1],CMA_SourceTarget[1],MI_SourceTarget[1]]\n",
    "\n",
    "#WT\n",
    "Total_Paths=pathCountHeatmap(G,Source_List,Target_List,'WildType')\n",
    "print(Total_Paths)\n",
    "#SQSTM1 KO\n",
    "KO_Total_Paths=geneKOHeatmap(['Q13501'],'SQSTM1 KO')\n",
    "print(KO_Total_Paths)\n",
    "#mTOR KO\n",
    "KO_Total_Paths=geneKOHeatmap(['P42345'],'mTOR KO')\n",
    "print(KO_Total_Paths)\n",
    "#Dynein motor\n",
    "KO_Total_Paths=geneKOHeatmap(['Q14204'],'Dynein motor KO')\n",
    "print(KO_Total_Paths)\n",
    "#BECN1\n",
    "KO_Total_Paths=geneKOHeatmap(['Q14457'],'BECLIN-1 KO')\n",
    "print(KO_Total_Paths)\n",
    "#AMBRA1\n",
    "KO_Total_Paths=geneKOHeatmap(['Q9C0C7'],'AMBRA1 KO')\n",
    "print(KO_Total_Paths)\n",
    "#ULK1\n",
    "KO_Total_Paths=geneKOHeatmap(['O75385'],'ULK1 KO')\n",
    "print(KO_Total_Paths)\n",
    "#PRKN\n",
    "KO_Total_Paths=geneKOHeatmap(['O60260'],'PARKIN KO')\n",
    "print(KO_Total_Paths)\n",
    "#LAMP2\n",
    "KO_Total_Paths=geneKOHeatmap(['P13473'],'LAMP2 KO')\n",
    "print(KO_Total_Paths)\n",
    "#EEF1A1\n",
    "KO_Total_Paths=geneKOHeatmap(['P68104'],'EEF1A1 KO')\n",
    "print(KO_Total_Paths)\n",
    "\n",
    "'''KO_Total_Paths={}\n",
    "for i in Protein_Dict:\n",
    "    k=i+' KO'\n",
    "    v=geneKOHeatmap([i],str(i)+' KO')\n",
    "    KO_Total_Paths[k]=v'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generating reaction-connection list for reviewer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating list for curatortool\n",
    "print(len(edgesPreced),len(edgesSharedPE),len(edgeDiff))\n",
    "edgeDiff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#code to generate report of connections for review\n",
    "import csv, re\n",
    "\n",
    "##function to read and extract files\n",
    "def fileReader(file):\n",
    "    with open(file,'r') as edgeCSV:\n",
    "        edgeReader = csv.reader(edgeCSV)\n",
    "        edges = [x for x in edgeReader][1:]\n",
    "    return edges\n",
    "\n",
    "##import files\n",
    "path='C:\\\\Users\\\\tvarusai\\\\EBI_Work\\\\Reactome\\\\Network Analysis Project\\\\Reaction_Connect_Project\\\\'\n",
    "edgeList = 'Autophagy_Reactions_Edges_PrecedEvent.csv'\n",
    "edgesPreced = fileReader(path+edgeList)\n",
    "edgeList='Autophagy_Reactions_Edges_SharedIO_SplitSets_SimpleEnityFalse_UbFalse.csv'\n",
    "edgesSharedPE = fileReader(path+edgeList)\n",
    "pathwayFile = 'autophagyReactionPathwayMap.csv'\n",
    "pathwayInfo = fileReader(path+pathwayFile)\n",
    "edgeDiff = [i for i in edgesPreced+edgesSharedPE if i not in edgesPreced]\n",
    "\n",
    "##manually added links that cannot be logically created\n",
    "hubPreced=['R-HSA-5205663','R-HSA-8959571','R-HSA-9646685','R-HSA-9613666','R-HSA-9664880','R-HSA-5682385']\n",
    "hubEvent='R-HSA-5679255'\n",
    "manuallyAddedLinks=[[rxn,hubEvent] for rxn in hubPreced]\n",
    "revEvents = edgeDiff + manuallyAddedLinks    #combining links generated logically with manual ones\n",
    "len(revEvents)\n",
    "\n",
    "##compiling linked events\n",
    "secondEvent=[event[1] for event in revEvents]\n",
    "secondEvent=set(secondEvent)\n",
    "print(len(revEvents),len(secondEvent))\n",
    "linkEvents=[]\n",
    "for item in secondEvent:\n",
    "    tempCont=[]\n",
    "    for event in revEvents:\n",
    "        if item == event[1]:\n",
    "            tempCont.append(event[0])\n",
    "    linkEvents.append([item,tempCont])\n",
    "linkEvents\n",
    "\n",
    "##converting reactions to hyperlinks with names\n",
    "###links in curator.reactome\n",
    "MA_ELV = 'https://curator.reactome.org/PathwayBrowser/#/R-HSA-1632852&SEL=R-HSA-5681999&PATH=R-HSA-9612973'\n",
    "SA_Mit_Pink_ELV = 'https://curator.reactome.org/PathwayBrowser/#/R-HSA-5205647&SEL=R-HSA-5205652&PATH=R-HSA-9612973,R-HSA-1632852,R-HSA-9663891'\n",
    "SA_Mit_Rec_ELV = 'https://curator.reactome.org/PathwayBrowser/#/R-HSA-5205647&SEL=R-HSA-8948136&PATH=R-HSA-9612973,R-HSA-1632852,R-HSA-9663891'\n",
    "SA_Agg_ELV = 'https://curator.reactome.org/PathwayBrowser/#/R-HSA-9646399&SEL=R-HSA-9641127&PATH=R-HSA-9612973,R-HSA-1632852,R-HSA-9663891'\n",
    "SA_Lip_ELV = 'https://curator.reactome.org/PathwayBrowser/#/R-HSA-9613354&SEL=R-HSA-9613565&PATH=R-HSA-9612973,R-HSA-1632852,R-HSA-9663891'\n",
    "SA_Pex_ELV = 'https://curator.reactome.org/PathwayBrowser/#/R-HSA-9664873&SEL=R-HSA-9664879&PATH=R-HSA-9612973,R-HSA-1632852,R-HSA-9663891'\n",
    "CMA_ELV = 'https://curator.reactome.org/PathwayBrowser/#/R-HSA-9613829&SEL=R-HSA-9622840&PATH=R-HSA-9612973'\n",
    "MI_ELV = 'https://curator.reactome.org/PathwayBrowser/#/R-HSA-9615710&SEL=R-HSA-9631068&PATH=R-HSA-9612973'\n",
    "###funtion to generate hyperlinks\n",
    "def linkGenerator(categ,rxnID,rxnNam):\n",
    "    if categ == 'Macroautophagy':\n",
    "        link = '=HYPERLINK(\"'+re.sub('SEL=R-HSA-\\w.*?&PATH','SEL='+rxnID+'&PATH',MA_ELV)+'\",\"'+rxnNam+'\")'\n",
    "        return link\n",
    "    if categ == 'Pink/Parkin Mediated Mitophagy':\n",
    "        link = '=HYPERLINK(\"'+re.sub('SEL=R-HSA-\\w.*?&PATH','SEL='+rxnID+'&PATH',SA_Mit_Pink_ELV)+'\",\"'+rxnNam+'\")'\n",
    "        return link\n",
    "    if categ == 'Receptor Mediated Mitophagy':\n",
    "        link = '=HYPERLINK(\"'+re.sub('SEL=R-HSA-\\w.*?&PATH','SEL='+rxnID+'&PATH',SA_Mit_Rec_ELV)+'\",\"'+rxnNam+'\")'\n",
    "        return link\n",
    "    if categ == 'Aggrephagy':\n",
    "        link = '=HYPERLINK(\"'+re.sub('SEL=R-HSA-\\w.*?&PATH','SEL='+rxnID+'&PATH',SA_Agg_ELV)+'\",\"'+rxnNam+'\")'\n",
    "        return link\n",
    "    if categ == 'Lipophagy':\n",
    "        link = '=HYPERLINK(\"'+re.sub('SEL=R-HSA-\\w.*?&PATH','SEL='+rxnID+'&PATH',SA_Lip_ELV)+'\",\"'+rxnNam+'\")'\n",
    "        return link\n",
    "    if categ == 'Pexophagy':\n",
    "        link = '=HYPERLINK(\"'+re.sub('SEL=R-HSA-\\w.*?&PATH','SEL='+rxnID+'&PATH',SA_Pex_ELV)+'\",\"'+rxnNam+'\")'\n",
    "        return link\n",
    "    if categ == 'Chaperone Mediated Autophagy':\n",
    "        link = '=HYPERLINK(\"'+re.sub('SEL=R-HSA-\\w.*?&PATH','SEL='+rxnID+'&PATH',CMA_ELV)+'\",\"'+rxnNam+'\")'\n",
    "        return link\n",
    "    if categ == 'Late endosomal microautophagy':\n",
    "        link = '=HYPERLINK(\"'+re.sub('SEL=R-HSA-\\w.*?&PATH','SEL='+rxnID+'&PATH',MI_ELV)+'\",\"'+rxnNam+'\")'\n",
    "        return link\n",
    "###creating the list of reactions\n",
    "for event in linkEvents:\n",
    "    for item in pathwayInfo:\n",
    "        if item[3] == event[0]:\n",
    "            patway = item[1]\n",
    "            rnID = item[3]\n",
    "            rnNam = item[2]\n",
    "    event[0] = linkGenerator(patway,rnID,rnNam)\n",
    "    lnks = []\n",
    "    for n in event[1]:\n",
    "        for item in pathwayInfo:\n",
    "            if item[3] == n:\n",
    "                patway = item[1]\n",
    "                rnID = item[3]\n",
    "                rnNam = item[2]\n",
    "#        print(patway,rnID,rnNam)\n",
    "        k = linkGenerator(patway,rnID,rnNam)\n",
    "        lnks.append(k)\n",
    "    event[1] = lnks\n",
    "    \n",
    "##writing files to csv\n",
    "with open(path+'autophagyLinksReview.csv','w',newline='') as file:\n",
    "    writer = csv.writer(file)\n",
    "    writer.writerows(linkEvents)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Logic Network Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cypher\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\"\n",
    "\n",
    "#run query for pathway name-ID map\n",
    "PathwayName_ID_CQ=\"\"\"\n",
    "MATCH (pa:TopLevelPathway{speciesName:\"Homo sapiens\"})\n",
    "RETURN pa.displayName AS PathwayNames, pa.stId AS PathwayStIds\n",
    "\"\"\" \n",
    "PathwayMap_DF=cypher.run(PathwayName_ID_CQ,conn=Con).get_dataframe()\n",
    "PathwayMap_Dict=PathwayMap_DF.set_index('PathwayNames')['PathwayStIds'].to_dict()\n",
    "\n",
    "m=0\n",
    "for name in PathwayMap_Dict:\n",
    "\n",
    "    Pathway_Name=name\n",
    "    Pathway_stId='\"'+PathwayMap_Dict[Pathway_Name]+'\"'\n",
    "\n",
    "    #run query to generate nodes and edges for logic network model\n",
    "    Logic_Nodes_CQ=\"\"\"\n",
    "    MATCH(tpa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    WITH ev\n",
    "    OPTIONAL MATCH(ev)-[:input]->(pin:PhysicalEntity)\n",
    "    OPTIONAL MATCH (pin)-[:hasMember|hasCandidate*]->(pins:PhysicalEntity)\n",
    "    WITH ev,pin,pins\n",
    "    OPTIONAL MATCH(ev)-[:output]->(pou:PhysicalEntity)\n",
    "    OPTIONAL MATCH (pou)-[:hasMember|hasCandidate*]->(pous:PhysicalEntity)\n",
    "    WITH ev,pin,pins,pou,pous\n",
    "    OPTIONAL MATCH(ev)-[:catalystActivity|physicalEntity*]->(pca:PhysicalEntity)\n",
    "    OPTIONAL MATCH(pca)-[:hasMember|hasCandidate*]->(pcas:PhysicalEntity)\n",
    "    WITH ev,pin,pins,pou,pous,pca,pcas\n",
    "    OPTIONAL MATCH(ev)-[:regulatedBy|physicalEntity*]->(reg:Regulation)-[:regulator]->(pre:PhysicalEntity)\n",
    "    OPTIONAL MATCH(pre)-[:hasMember|hasCandidate*]->(pres:PhysicalEntity)\n",
    "    WITH ev, reg, \n",
    "    COLLECT(pin.stId)+COLLECT(pins.stId) AS i, COLLECT(pou.stId)+COLLECT(pous.stId) AS o, COLLECT(pca.stId)+COLLECT(pcas.stId) AS c, COLLECT(pre.stId)+COLLECT(pres.stId) AS r\n",
    "    RETURN DISTINCT ev.stId AS Reaction, i AS Input, o AS Output, c as Catalyst, r AS Regulator, reg.displayName AS Regulation\n",
    "    \"\"\"\n",
    "    Logic_Edges_CQ=\"\"\"\n",
    "    //find reaction connections for all except Set connectors\n",
    "    ///remove drug-related reactions\n",
    "    MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "    WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "    WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "    ///check for entity sharing after filtering for simple entity and Ub\n",
    "    WITH pa, ro\n",
    "    MATCH(ro)-[:output]->(po1:PhysicalEntity)\n",
    "    WHERE NOT (po1.schemaClass=\"SimpleEntity\" OR po1.stId=\"R-HSA-113595\" OR po1.schemaClass=\"DefinedSet\" OR po1.schemaClass=\"CandidateSet\")\n",
    "    WITH pa, ro, po1\n",
    "    MATCH(pa)-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc2:PhysicalEntity)\n",
    "    WITH  pa, ro, po1, ri1, COLLECT(DISTINCT dc2.schemaClass) AS alldc2\n",
    "    WHERE NOT (\"ChemicalDrug\" IN (alldc2) OR \"ProteinDrug\" IN (alldc2))\n",
    "    WITH ro, po1, ri1\n",
    "    MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "    OPTIONAL MATCH (cacomp:Compartment)<-[:compartment]-(po1)<-[:catalystActivity|physicalEntity*]-(ri1)\n",
    "    OPTIONAL MATCH (recomp:Compartment)<-[:compartment]-(po1)<-[:regulator|physicalEntity*]-(reg:Regulation)<-[:regulatedBy|physicalEntity*]-(ri1)\n",
    "    WITH ro, ri1, po1, reg, cacomp, recomp\n",
    "    RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction, po1.schemaClass AS SharedEntityClass, po1.stId AS SharedEntity\n",
    "    ORDER BY ro.stId\n",
    "\n",
    "    //find reaction connections for set connectors\n",
    "    ///remove drug-related reactions\n",
    "    UNION MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "    WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "    WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "    ///check for entity sharing after filtering for simple entity and Ub\n",
    "    WITH pa, ro\n",
    "    MATCH(ro)-[:output]->(po2:PhysicalEntity)-[:hasMember|hasCandidate*]->(po3:PhysicalEntity)\n",
    "    WHERE NOT (po3.stId=\"R-HSA-113595\") AND (po2.schemaClass=\"DefinedSet\" OR po2.schemaClass=\"CandidateSet\")\n",
    "    WITH pa, ro, po3\n",
    "    MATCH(pa)-[:hasEvent*]->(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2)\n",
    "    OPTIONAL MATCH (cacomp:Compartment)<-[:compartment]-(po3)<-[:catalystActivity|physicalEntity*]-(ri2)\n",
    "    OPTIONAL MATCH (recomp:Compartment)<-[:compartment]-(po3)<-[:regulator|physicalEntity*]-(reg:Regulation)<-[:regulatedBy|physicalEntity*]-(ri2)\n",
    "    WITH ro, ri2, po3, reg, cacomp, recomp\n",
    "    RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction, po3.schemaClass AS SharedEntityClass, po3.stId AS SharedEntity\n",
    "    ORDER BY ro.stId\n",
    "    \"\"\"\n",
    "\n",
    "    Logic_Nodes_DF=cypher.run(Logic_Nodes_CQ,conn=Con).get_dataframe()\n",
    "    Logic_Edges_DF=cypher.run(Logic_Edges_CQ,conn=Con).get_dataframe()\n",
    "\n",
    "    #data processing\n",
    "    ##creating logic rules and induction/supression value for edges\n",
    "    BoolEdgesList=[]\n",
    "    Input_List=[y for x in list(Logic_Nodes_DF['Input']) for y in x]\n",
    "    InputCounts={i:Input_List.count(i) for i in set(Input_List)}\n",
    "    for label,row in Logic_Edges_DF.iterrows():\n",
    "        SharedEntity=row['SharedEntity']\n",
    "        #label shared entity\n",
    "        if any(SharedEntity in x for x in list(Logic_Nodes_DF['Input'])):\n",
    "            SharedType='Input'\n",
    "        elif any(SharedEntity in x for x in list(Logic_Nodes_DF['Catalyst'])):\n",
    "            SharedType='Catalyst'\n",
    "        elif any(SharedEntity in x for x in list(Logic_Nodes_DF['Regulator'])):\n",
    "            RegulationType=[Logic_Nodes_DF.iloc[i]['Regulation'][:8] for i,v in Logic_Nodes_DF[['Regulator']].iterrows() if SharedEntity in v['Regulator']][0]\n",
    "            SharedType=RegulationType+' Regulator'\n",
    "        else:\n",
    "            print(\"Shared entity not found!\")\n",
    "        #assign Boolean logic and activation/suppression \n",
    "        if SharedType=='Input':\n",
    "            if InputCounts[SharedEntity]==1:\n",
    "                Logic='AND'\n",
    "                Value=1\n",
    "            elif InputCounts[SharedEntity]>1:\n",
    "                Logic='OR'\n",
    "                Value=0\n",
    "        elif SharedType!='Input':\n",
    "            Logic='AND'\n",
    "            if 'Positive' in SharedType:\n",
    "                Value=1\n",
    "            elif 'Negative' in SharedType:\n",
    "                Value=-1\n",
    "            elif 'Catalyst' in SharedType:\n",
    "                Value=1\n",
    "        BoolEdgesList.append([row['First_Reaction'],row['Second_Reaction'],SharedEntity,SharedType,Logic,Value])\n",
    "    BoolEdges=pd.DataFrame(BoolEdgesList,columns=['First_Reaction','Second_Reaction','Shared_Entity','Shared_Entity_Type','Logic','Value'])\n",
    "    m+=1\n",
    "    print('Pathway Name: ',name,' Progress: ', m,'/',len(PathwayMap_Dict),' Network size: ',BoolEdges.shape[0])\n",
    "    #print('Network size',BoolEdges.shape)\n",
    "    BoolEdges.head()\n",
    "    #print('Before: ',BoolEdges.shape[0],' After: ',BoolEdges.drop_duplicates().shape[0])\n",
    "\n",
    "#     notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "#     path=notebook_path.rsplit('/',1)\n",
    "#     path=path[0]+'/Reaction_Logic/Boolean_Networks/'\n",
    "#     BoolEdges.to_csv(path+Pathway_Name+'_LogicNetwork.csv',index=False,header=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set(list(BoolEdges[BoolEdges['Shared_Entity_Type']=='Negative Regulator']['Value']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import networkx as nx\n",
    "A = np.array([[1, 1], [2, 1]])\n",
    "#G = nx.from_numpy_matrix(A)\n",
    "G = nx.from_numpy_matrix(A, create_using=nx.MultiGraph)\n",
    "v=nx.spring_layout(G)\n",
    "nx.draw(G,v,node_color='y')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Neo4j cypher queries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cypher\n",
    "\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\"\n",
    "cq=\"\"\"\n",
    "MATCH (pa:TopLevelPathway{speciesName:'Homo sapiens'})-[:hasEvent*]->(re:ReactionLikeEvent)\n",
    "RETURN DISTINCT pa.displayName AS PathwayName, COUNT(DISTINCT re.stId) AS ReactionCount\n",
    "\"\"\"\n",
    "cq=\"\"\"\n",
    "MATCH (pa:TopLevelPathway{speciesName:'Homo sapiens'})-[:hasEvent*]->(re:ReactionLikeEvent)\n",
    "RETURN DISTINCT pa.displayName AS PathwayName, COUNT(DISTINCT re.stId) AS ReactionCount\n",
    "\"\"\"\n",
    "\n",
    "results=cypher.run(cq,conn=Con).get_dataframe()\n",
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pathway connectivity analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import cypher\n",
    "import os\n",
    "\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\" #database connection\n",
    "\n",
    "#run query for pathway name-ID map\n",
    "PathwayName_ID_CQ=\"\"\"\n",
    "MATCH (pa:TopLevelPathway{speciesName:\"Homo sapiens\"})\n",
    "RETURN pa.displayName AS PathwayNames, pa.stId AS PathwayStIds\n",
    "\"\"\" \n",
    "PathwayMap_DF=cypher.run(PathwayName_ID_CQ,conn=Con).get_dataframe()\n",
    "PathwayMap_Dict=PathwayMap_DF.set_index('PathwayNames')['PathwayStIds'].to_dict()\n",
    "\n",
    "\n",
    "#run query for event-author map\n",
    "Event_Author_CQ=\"\"\"\n",
    "MATCH(pa:TopLevelPathway)-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "OPTIONAL MATCH (ev)-[:authored]-(ie:InstanceEdit)-[:author]-(au:Person)\n",
    "RETURN DISTINCT ev.stId AS EventID, au.displayName AS Author\n",
    "\"\"\"\n",
    "EventMap_DF=cypher.run(Event_Author_CQ,conn=Con).get_dataframe()\n",
    "EventMap_Dict=EventMap_DF.set_index('EventID')['Author'].to_dict()\n",
    "\n",
    "\n",
    "# #run query for SBGN-reaction map\n",
    "# SBGN_Event_CQ=\"\"\"\n",
    "# MATCH (pa:Pathway{speciesName:\"Homo sapiens\"})-[:hasEvent]->(ev:Event)\n",
    "# WHERE (pa.hasDiagram=true) AND NOT EXISTS (pa.hasEHLD)// AND NOT (ev.schemaClass='Pathway')\n",
    "# RETURN DISTINCT pa.stId AS PathwayID,  pa.displayName AS PathwayName, ev.stId AS EventID\n",
    "# \"\"\"\n",
    "# SBGNMap_DF=cypher.run(SBGN_Event_CQ,conn=Con).get_dataframe()\n",
    "# SBGN_IDMap_Dict=SBGNMap_DF.set_index('EventID')['PathwayID'].to_dict()\n",
    "# SBGN_NameMap_Dict=SBGNMap_DF.set_index('EventID')['PathwayName'].to_dict()\n",
    "\n",
    "\n",
    "#run query for reaction ID-name map\n",
    "ReactionIDName_CQ=\"\"\"\n",
    "MATCH (re:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "RETURN DISTINCT re.stId AS ReactionID, re.displayName AS ReactionName\n",
    "\"\"\"\n",
    "ReacMap_DF=cypher.run(ReactionIDName_CQ,conn=Con).get_dataframe()\n",
    "ReacMap_Dict=ReacMap_DF.set_index('ReactionID')['ReactionName'].to_dict()\n",
    "\n",
    "\n",
    "#Analyze missing connections for each top-level pathway\n",
    "for i in sorted(PathwayMap_Dict):\n",
    "#for i in ['Cell Cycle']:\n",
    "#     # skip big pathways to avoid memory issues\n",
    "#     if i in ['Immune System','Signal Transduction']:\n",
    "#         continue\n",
    "    \n",
    "    print(i)\n",
    "    \n",
    "    #assign pathway ID\n",
    "    Pathway_Name=i\n",
    "    Pathway_stId='\"'+PathwayMap_Dict[i]+'\"'\n",
    "    \n",
    "    #run query to get precedingEvents connections\n",
    "    Preced_CQ=\"\"\"\n",
    "    MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    MATCH(ev)-[:precedingEvent]->(pe:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    RETURN DISTINCT pe.stId AS First_Reaction, ev.stId AS Second_Reaction\n",
    "    \"\"\"\n",
    "    Preced_DF=cypher.run(Preced_CQ,conn=Con).get_dataframe()\n",
    "    \n",
    "    if i=='Metabolism':\n",
    "        #query for shared entity connections in Metabolism factoring in simple entities\n",
    "        Shared_CQ=\"\"\"\n",
    "        MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "        MATCH(ro)-[:input|output|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate]->(po1:PhysicalEntity)-[:referenceEntity]->(ref0:ReferenceMolecule)\n",
    "        WHERE NOT ANY(x IN ref0.identifier WHERE x IN [\"15377\",\"16761\",\"15422\",\"16027\",\"15996\",\"17552\",\"17659\",\"15713\",\"15378\",\"16234\",\"18367\",\"29888\",\"13392\",\"18009\",\"16908\",\"16238\",\"17877\",\"15346\",\"15379\",\"16526\",\"15846\",\"17345\",\"16174\",\"17713\",\"16284\",\"28862\",\"16192\",\"16497\",\"17625\",\"15713\",\"16695\",\"18009\"])\n",
    "        WITH pa, ro, po1\n",
    "        MATCH(po1)<-[:output]-(ro)\n",
    "        WHERE NOT (po1.stId=\"R-HSA-113595\")\n",
    "        WITH pa, ro, po1\n",
    "        MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity]-(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)\n",
    "        WITH ro, ri1\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction\n",
    "        UNION MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2:PhysicalEntity{schemaClass:\"DefinedSet\"})-[:output|physicalEntity|hasMember|hasCandidate]->(po3:PhysicalEntity)\n",
    "        WHERE NOT (po2.stId=\"R-HSA-113595\")\n",
    "        WITH pa, ro, po3\n",
    "        MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate]-(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)\n",
    "        WITH ro, ri2\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction\n",
    "        \"\"\"\n",
    "    else:\n",
    "        #query for shared entity connections\n",
    "        Shared_CQ=\"\"\"\n",
    "        MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "        WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "        WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "        WITH pa, ro\n",
    "        MATCH(ro)-[:output]->(po1:PhysicalEntity)\n",
    "        WHERE NOT (po1.schemaClass=\"SimpleEntity\" OR po1.stId=\"R-HSA-113595\")\n",
    "        WITH pa, ro, po1\n",
    "        MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc2:PhysicalEntity)\n",
    "        WITH  pa, ro, po1, ri1, COLLECT(DISTINCT dc2.schemaClass) AS alldc2\n",
    "        WHERE NOT (\"ChemicalDrug\" IN (alldc2) OR \"ProteinDrug\" IN (alldc2))\n",
    "        WITH ro, po1, ri1\n",
    "        MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "        WITH ro, ri1\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction\n",
    "        ORDER BY ro.stId\n",
    "        UNION MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2:PhysicalEntity{schemaClass:\"DefinedSet\"})-[:output|physicalEntity|hasMember|hasCandidate*]->(po3:PhysicalEntity)\n",
    "        WHERE NOT (po2.stId=\"R-HSA-113595\")\n",
    "        WITH pa, ro, po3\n",
    "        MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)\n",
    "        WITH ro, ri2\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction\n",
    "        ORDER BY ro.stId\n",
    "        \"\"\"\n",
    "        \n",
    "    #run query to get shared entity connections\n",
    "    Shared_DF=cypher.run(Shared_CQ,conn=Con).get_dataframe()\n",
    "    \n",
    "    # find missing reaction connections \n",
    "    Diff_DF=Shared_DF[~Shared_DF.apply(tuple,1).isin(Preced_DF.apply(tuple,1))]\n",
    "    print('# missing connections = '+str(len(Diff_DF)))\n",
    "\n",
    "    # generate hyperlinks\n",
    "    Out_DF=Diff_DF\n",
    "    #Out_DF['SBGN Diagram']=Diff_DF.Second_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/PathwayBrowser/#/'+str(SBGN_IDMap_Dict[s])+'\",\"'+str(SBGN_NameMap_Dict[s])+'\")')\n",
    "    Out_DF['First Reaction']=Diff_DF.First_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(ReacMap_Dict[s])+'\")')\n",
    "    Out_DF['Second Reaction']=Diff_DF.Second_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(ReacMap_Dict[s])+'\")')\n",
    "    Out_DF['Second Reaction Author']=Diff_DF.Second_Reaction.apply(lambda a: EventMap_Dict[a])\n",
    "    #Out_DF.drop('First_Reaction',inplace=True,axis=1)\n",
    "    #Out_DF.drop('Second_Reaction',inplace=True,axis=1)\n",
    "\n",
    "    # write to file \n",
    "    ## path\n",
    "    notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "    path=notebook_path.rsplit('/',1)\n",
    "    path=path[0]+'/Reaction_Connectivity/Pathways/'\n",
    "    Out_DF.to_csv(path+'Edited_'+Pathway_Name+'_MissingConnections.csv',index=False,header=True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MATCH(pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "WITH pa, ro\n",
    "MATCH(ro)-[:output]->(po1:PhysicalEntity)\n",
    "WHERE NOT (po1.schemaClass=\"SimpleEntity\" OR po1.stId=\"R-HSA-113595\")\n",
    "WITH pa, ro, po1\n",
    "MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc2:PhysicalEntity)\n",
    "WITH  pa, ro, po1, ri1, COLLECT(DISTINCT dc2.schemaClass) AS alldc2\n",
    "WHERE NOT (\"ChemicalDrug\" IN (alldc2) OR \"ProteinDrug\" IN (alldc2))\n",
    "WITH ro, po1, ri1\n",
    "MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "WITH ro, ri1\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction\n",
    "ORDER BY ro.stId\n",
    "UNION MATCH(pa:Pathway{stId:\"R-HSA-9612973\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2:PhysicalEntity{schemaClass:\"DefinedSet\"})-[:output|physicalEntity|hasMember|hasCandidate*]->(po3:PhysicalEntity)\n",
    "WHERE NOT (po2.stId=\"R-HSA-113595\")\n",
    "WITH pa, ro, po3\n",
    "MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)\n",
    "WITH ro, ri2\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction\n",
    "ORDER BY ro.stId"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27 rows affected.\n",
      "14371 rows affected.\n",
      "13345 rows affected.\n",
      "Metabolism\n",
      "2391 rows affected.\n",
      "69879 rows affected.\n",
      "# missing connections = 50871\n"
     ]
    }
   ],
   "source": [
    "#improving the algorithm to filter false negatives\n",
    "\n",
    "import pandas as pd\n",
    "import cypher\n",
    "import os\n",
    "\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\" #database connection\n",
    "\n",
    "#run query for pathway name-ID map\n",
    "PathwayName_ID_CQ=\"\"\"\n",
    "MATCH (pa:TopLevelPathway{speciesName:\"Homo sapiens\"})\n",
    "RETURN pa.displayName AS PathwayNames, pa.stId AS PathwayStIds\n",
    "\"\"\" \n",
    "PathwayMap_DF=cypher.run(PathwayName_ID_CQ,conn=Con).get_dataframe()\n",
    "PathwayMap_Dict=PathwayMap_DF.set_index('PathwayNames')['PathwayStIds'].to_dict()\n",
    "\n",
    "\n",
    "#run query for event-author map\n",
    "Event_Author_CQ=\"\"\"\n",
    "MATCH(pa:TopLevelPathway)-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "OPTIONAL MATCH (ev)-[:authored]-(ie:InstanceEdit)-[:author]-(au:Person)\n",
    "RETURN DISTINCT ev.stId AS EventID, au.displayName AS Author\n",
    "\"\"\"\n",
    "EventMap_DF=cypher.run(Event_Author_CQ,conn=Con).get_dataframe()\n",
    "EventMap_Dict=EventMap_DF.set_index('EventID')['Author'].to_dict()\n",
    "\n",
    "\n",
    "# #run query for SBGN-reaction map\n",
    "# SBGN_Event_CQ=\"\"\"\n",
    "# MATCH (pa:Pathway{speciesName:\"Homo sapiens\"})-[:hasEvent]->(ev:Event)\n",
    "# WHERE (pa.hasDiagram=true) AND NOT EXISTS (pa.hasEHLD)// AND NOT (ev.schemaClass='Pathway')\n",
    "# RETURN DISTINCT pa.stId AS PathwayID,  pa.displayName AS PathwayName, ev.stId AS EventID\n",
    "# \"\"\"\n",
    "# SBGNMap_DF=cypher.run(SBGN_Event_CQ,conn=Con).get_dataframe()\n",
    "# SBGN_IDMap_Dict=SBGNMap_DF.set_index('EventID')['PathwayID'].to_dict()\n",
    "# SBGN_NameMap_Dict=SBGNMap_DF.set_index('EventID')['PathwayName'].to_dict()\n",
    "\n",
    "\n",
    "#run query for reaction ID-name map\n",
    "ReactionIDName_CQ=\"\"\"\n",
    "MATCH (re:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "RETURN DISTINCT re.stId AS ReactionID, re.displayName AS ReactionName\n",
    "\"\"\"\n",
    "ReacMap_DF=cypher.run(ReactionIDName_CQ,conn=Con).get_dataframe()\n",
    "ReacMap_Dict=ReacMap_DF.set_index('ReactionID')['ReactionName'].to_dict()\n",
    "\n",
    "#Analyze missing connections for each top-level pathway\n",
    "#for i in sorted(PathwayMap_Dict):\n",
    "for i in ['Metabolism']:\n",
    "#     # skip big pathways to avoid memory issues\n",
    "#     if i in ['Immune System','Signal Transduction']:\n",
    "#         continue\n",
    "    \n",
    "    print(i)\n",
    "    \n",
    "    #assign pathway ID\n",
    "    Pathway_Name=i\n",
    "    Pathway_stId='\"'+PathwayMap_Dict[i]+'\"'\n",
    "    \n",
    "    #run query to get precedingEvents connections\n",
    "    Preced_CQ=\"\"\"\n",
    "    MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    MATCH(ev)-[:precedingEvent]->(pe:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    RETURN DISTINCT pe.stId AS First_Reaction, ev.stId AS Second_Reaction\n",
    "    \"\"\"\n",
    "    Preced_DF=cypher.run(Preced_CQ,conn=Con).get_dataframe()\n",
    "    \n",
    "    if i=='Metabolism':\n",
    "        ## path\n",
    "        notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "        path=notebook_path.rsplit('/',1)\n",
    "        path=path[0]+'/Reaction_Connectivity/'\n",
    "        Met_Triv_Mol=pd.read_excel(path+'Metabolism_TrivialChemicals.xlsx')\n",
    "        Exclude_List=list(map(str,list(Met_Triv_Mol['identifier'])))\n",
    "        #query for shared entity connections in Metabolism factoring in simple entities\n",
    "        Shared_CQ=\"\"\"\n",
    "        //find reaction connections for all except Set connectors\n",
    "        ///remove metabolism-specific exclude list\n",
    "        MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "        MATCH(ro)-[:input|output|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate]->(po0:PhysicalEntity)-[:referenceEntity]->(ref0:ReferenceMolecule)\n",
    "        WHERE NOT ANY(x IN ref0.identifier WHERE x IN \"\"\"+str(Exclude_List)+\"\"\") \n",
    "        ///remove drug-related reactions\n",
    "        WITH pa,ro\n",
    "        MATCH(ro)-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "        WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "        WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "        ///check for entity sharing after filtering for simple entity and Ub\n",
    "        WITH pa, ro\n",
    "        MATCH(ro)-[:output]->(po1:PhysicalEntity)\n",
    "        //WHERE NOT (po1.schemaClass=\"SimpleEntity\" OR po1.stId=\"R-HSA-113595\" OR po1.schemaClass=\"DefinedSet\" OR po1.schemaClass=\"CandidateSet\")\n",
    "        WHERE NOT (po1.stId=\"R-HSA-113595\" OR po1.schemaClass=\"DefinedSet\" OR po1.schemaClass=\"CandidateSet\")\n",
    "        WITH pa, ro, po1\n",
    "        MATCH(pa)-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc2:PhysicalEntity)\n",
    "        WITH  pa, ro, po1, ri1, COLLECT(DISTINCT dc2.schemaClass) AS alldc2\n",
    "        WHERE NOT (\"ChemicalDrug\" IN (alldc2) OR \"ProteinDrug\" IN (alldc2))\n",
    "        WITH ro, po1, ri1\n",
    "        MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "        OPTIONAL MATCH (cacomp:Compartment)<-[:compartment]-(po1)<-[:catalystActivity|physicalEntity*]-(ri1)\n",
    "        OPTIONAL MATCH (recomp:Compartment)<-[:compartment]-(po1)<-[:regulator|physicalEntity*]-(reg:Regulation)<-[:regulatedBy|physicalEntity*]-(ri1)\n",
    "        WITH ro, ri1, po1, reg, cacomp, recomp\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction, po1.schemaClass AS SharedEntityClass, po1.stId AS SharedEntity\n",
    "        ORDER BY ro.stId\n",
    "\n",
    "        //find reaction connections for set connectors\n",
    "        ///remove drug-related reactions\n",
    "        UNION MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "        WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "        WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "        ///check for entity sharing after filtering for simple entity and Ub\n",
    "        WITH pa, ro\n",
    "        MATCH(ro)-[:output]->(po2:PhysicalEntity)-[:hasMember|hasCandidate*]->(po3:PhysicalEntity)\n",
    "        WHERE NOT (po3.stId=\"R-HSA-113595\") AND (po2.schemaClass=\"DefinedSet\" OR po2.schemaClass=\"CandidateSet\")\n",
    "        WITH pa, ro, po3\n",
    "        MATCH(pa)-[:hasEvent*]->(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "        MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2)\n",
    "        OPTIONAL MATCH (cacomp:Compartment)<-[:compartment]-(po3)<-[:catalystActivity|physicalEntity*]-(ri2)\n",
    "        OPTIONAL MATCH (recomp:Compartment)<-[:compartment]-(po3)<-[:regulator|physicalEntity*]-(reg:Regulation)<-[:regulatedBy|physicalEntity*]-(ri2)\n",
    "        WITH ro, ri2, po3, reg, cacomp, recomp\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction, po3.schemaClass AS SharedEntityClass, po3.stId AS SharedEntity\n",
    "        ORDER BY ro.stId\n",
    "        \"\"\"\n",
    "    else:\n",
    "        #query for shared entity connections\n",
    "        Shared_CQ=\"\"\"\n",
    "        //find reaction connections for all except Set connectors\n",
    "        ///remove drug-related reactions\n",
    "        MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "        WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "        WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "        ///check for entity sharing after filtering for simple entity and Ub\n",
    "        WITH pa, ro\n",
    "        MATCH(ro)-[:output]->(po1:PhysicalEntity)\n",
    "        WHERE NOT (po1.schemaClass=\"SimpleEntity\" OR po1.stId=\"R-HSA-113595\" OR po1.schemaClass=\"DefinedSet\" OR po1.schemaClass=\"CandidateSet\")\n",
    "        WITH pa, ro, po1\n",
    "        MATCH(pa)-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc2:PhysicalEntity)\n",
    "        WITH  pa, ro, po1, ri1, COLLECT(DISTINCT dc2.schemaClass) AS alldc2\n",
    "        WHERE NOT (\"ChemicalDrug\" IN (alldc2) OR \"ProteinDrug\" IN (alldc2))\n",
    "        WITH ro, po1, ri1\n",
    "        MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "        OPTIONAL MATCH (cacomp:Compartment)<-[:compartment]-(po1)<-[:catalystActivity|physicalEntity*]-(ri1)\n",
    "        OPTIONAL MATCH (recomp:Compartment)<-[:compartment]-(po1)<-[:regulator|physicalEntity*]-(reg:Regulation)<-[:regulatedBy|physicalEntity*]-(ri1)\n",
    "        WITH ro, ri1, po1, reg, cacomp, recomp\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction, po1.schemaClass AS SharedEntityClass, po1.stId AS SharedEntity\n",
    "        ORDER BY ro.stId\n",
    "\n",
    "        //find reaction connections for set connectors\n",
    "        ///remove drug-related reactions\n",
    "        UNION MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "        WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "        WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "        ///check for entity sharing after filtering for simple entity and Ub\n",
    "        WITH pa, ro\n",
    "        MATCH(ro)-[:output]->(po2:PhysicalEntity)-[:hasMember|hasCandidate*]->(po3:PhysicalEntity)\n",
    "        WHERE NOT (po3.stId=\"R-HSA-113595\") AND (po2.schemaClass=\"DefinedSet\" OR po2.schemaClass=\"CandidateSet\")\n",
    "        WITH pa, ro, po3\n",
    "        MATCH(pa)-[:hasEvent*]->(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "        MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2)\n",
    "        OPTIONAL MATCH (cacomp:Compartment)<-[:compartment]-(po3)<-[:catalystActivity|physicalEntity*]-(ri2)\n",
    "        OPTIONAL MATCH (recomp:Compartment)<-[:compartment]-(po3)<-[:regulator|physicalEntity*]-(reg:Regulation)<-[:regulatedBy|physicalEntity*]-(ri2)\n",
    "        WITH ro, ri2, po3, reg, cacomp, recomp\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction, po3.schemaClass AS SharedEntityClass, po3.stId AS SharedEntity\n",
    "        ORDER BY ro.stId\n",
    "        \"\"\"\n",
    "        \n",
    "    #run query to get shared entity connections\n",
    "    Shared_DF=cypher.run(Shared_CQ,conn=Con).get_dataframe()\n",
    "    \n",
    "    #find missing reaction connections \n",
    "    Diff_DF=Shared_DF[~Shared_DF.apply(tuple,1).isin(Preced_DF.apply(tuple,1))]\n",
    "    \n",
    "    #remove cyclic reactions\n",
    "    Index_List=[]\n",
    "    com1=Diff_DF['First_Reaction']+'---'+Diff_DF['Second_Reaction'] #creating reaction pair\n",
    "    com2=Diff_DF['Second_Reaction']+'---'+Diff_DF['First_Reaction'] #creating reverse reaction pair\n",
    "    #check if reaction pair equals reverse reaction pair\n",
    "    for ind,val in enumerate(com1):\n",
    "        if val in list(com2):\n",
    "            Index_List.append(ind)\n",
    "    CycNull_DF=Diff_DF.drop(Diff_DF.index[Index_List])\n",
    "    \n",
    "    print('# missing connections = '+str(len(CycNull_DF)))\n",
    "\n",
    "    # generate hyperlinks\n",
    "    Out_DF=CycNull_DF\n",
    "    #Out_DF['SBGN Diagram']=Diff_DF.Second_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/PathwayBrowser/#/'+str(SBGN_IDMap_Dict[s])+'\",\"'+str(SBGN_NameMap_Dict[s])+'\")')\n",
    "    Out_DF['First Reaction']=CycNull_DF.First_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(ReacMap_Dict[s])+'\")')\n",
    "    Out_DF['Second Reaction']=CycNull_DF.Second_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(ReacMap_Dict[s])+'\")')\n",
    "    Out_DF['Second Reaction Author']=CycNull_DF.Second_Reaction.apply(lambda a: EventMap_Dict[a])\n",
    "    #Out_DF.drop('First_Reaction',inplace=True,axis=1)\n",
    "    #Out_DF.drop('Second_Reaction',inplace=True,axis=1)\n",
    "\n",
    "#     # write to file \n",
    "#     ## path\n",
    "#     notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "#     path=notebook_path.rsplit('/',1)\n",
    "#     path=path[0]+'/Reaction_Connectivity/Pathways/'\n",
    "#     Out_DF.to_csv(path+'2Edited_'+Pathway_Name+'_MissingConnections.csv',index=False,header=True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import cypher\n",
    "import os\n",
    "\n",
    "Pathway_stId = '\"'+'R-HSA-9612973'+'\"'\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\" #database connection\n",
    "\n",
    "## path\n",
    "notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "path=notebook_path.rsplit('/',1)\n",
    "path=path[0]+'/Reaction_Connectivity/'\n",
    "Met_Triv_Mol=pd.read_excel(path+'Metabolism_TrivialChemicals.xlsx')\n",
    "Exclude_List=list(map(str,list(Met_Triv_Mol['identifier'])))\n",
    "\n",
    "#query for shared entity connections in Metabolism factoring in simple entities\n",
    "Shared_CQ=\"\"\"\n",
    "MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "MATCH(ro)-[:input|output|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate]->(po1:PhysicalEntity)-[:referenceEntity]->(ref0:ReferenceMolecule)\n",
    "WHERE NOT ANY(x IN ref0.identifier WHERE x IN \"\"\"+str(Exclude_List)+\"\"\") WITH pa, ro, po1\n",
    "MATCH(po1)<-[:output]-(ro)\n",
    "WHERE NOT (po1.stId=\"R-HSA-113595\")\n",
    "WITH pa, ro, po1\n",
    "MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity]-(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)\n",
    "WITH ro, ri1\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction\n",
    "UNION MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2:PhysicalEntity{schemaClass:\"DefinedSet\"})-[:output|physicalEntity|hasMember|hasCandidate]->(po3:PhysicalEntity)\n",
    "WHERE NOT (po2.stId=\"R-HSA-113595\")\n",
    "WITH pa, ro, po3\n",
    "MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate]-(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)\n",
    "WITH ro, ri2\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction\n",
    "\"\"\"\n",
    "Shared_DF=cypher.run(Shared_CQ,conn=Con).get_dataframe()\n",
    "Shared_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# connectivity for all of Reactome\n",
    "\n",
    "import pandas as pd\n",
    "import cypher\n",
    "import os\n",
    "\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\" #database connection\n",
    "\n",
    "#run query for pathway name-ID map\n",
    "PathwayName_ID_CQ=\"\"\"\n",
    "MATCH (pa:TopLevelPathway{speciesName:\"Homo sapiens\"})\n",
    "RETURN pa.displayName AS PathwayNames, pa.stId AS PathwayStIds\n",
    "\"\"\" \n",
    "PathwayMap_DF=cypher.run(PathwayName_ID_CQ,conn=Con).get_dataframe()\n",
    "PathwayMap_Dict=PathwayMap_DF.set_index('PathwayNames')['PathwayStIds'].to_dict()\n",
    "\n",
    "\n",
    "#run query for event-author map\n",
    "Event_Author_CQ=\"\"\"\n",
    "MATCH(pa:TopLevelPathway)-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "OPTIONAL MATCH (ev)-[:authored]-(ie:InstanceEdit)-[:author]-(au:Person)\n",
    "RETURN DISTINCT ev.stId AS EventID, au.displayName AS Author\n",
    "\"\"\"\n",
    "EventMap_DF=cypher.run(Event_Author_CQ,conn=Con).get_dataframe()\n",
    "EventMap_Dict=EventMap_DF.set_index('EventID')['Author'].to_dict()\n",
    "\n",
    "#run query to get precedingEvents connections\n",
    "Preced_CQ=\"\"\"\n",
    "MATCH(pa:Pathway)-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"}) \n",
    "MATCH(ev)-[:precedingEvent]->(pe:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "RETURN DISTINCT pe.stId AS First_Reaction, ev.stId AS Second_Reaction\n",
    "\"\"\"\n",
    "Preced_DF=cypher.run(Preced_CQ,conn=Con).get_dataframe()\n",
    "\n",
    "#run query to get shared entity connections\n",
    "Shared_CQ=\"\"\"\n",
    "MATCH(pa:Pathway)-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "WITH pa, ro\n",
    "MATCH(ro)-[:output]->(po1:PhysicalEntity)\n",
    "WHERE NOT (po1.schemaClass=\"SimpleEntity\" OR po1.stId=\"R-HSA-113595\")\n",
    "WITH pa, ro, po1\n",
    "MATCH(pa:Pathway)-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc2:PhysicalEntity)\n",
    "WITH  pa, ro, po1, ri1, COLLECT(DISTINCT dc2.schemaClass) AS alldc2\n",
    "WHERE NOT (\"ChemicalDrug\" IN (alldc2) OR \"ProteinDrug\" IN (alldc2))\n",
    "WITH ro, po1, ri1\n",
    "MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "WITH ro, ri1\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction\n",
    "ORDER BY ro.stId\n",
    "UNION MATCH(pa:Pathway)-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2:PhysicalEntity{schemaClass:\"DefinedSet\"})-[:output|physicalEntity|hasMember|hasCandidate*]->(po3:PhysicalEntity)\n",
    "WHERE NOT (po2.stId=\"R-HSA-113595\")\n",
    "WITH pa, ro, po3\n",
    "MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)\n",
    "WITH ro, ri2\n",
    "RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction\n",
    "ORDER BY ro.stId\n",
    "\"\"\"\n",
    "Shared_DF=cypher.run(Shared_CQ,conn=Con).get_dataframe()\n",
    "\n",
    "# find missing reaction connections \n",
    "Diff_DF=Shared_DF[~Shared_DF.apply(tuple,1).isin(Preced_DF.apply(tuple,1))]\n",
    "print('# missing connections = '+str(len(Diff_DF)))\n",
    "\n",
    "# generate hyperlinks\n",
    "Out_DF=Diff_DF\n",
    "Out_DF['First Reaction']=Diff_DF.First_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(s)+'\")')\n",
    "Out_DF['Second Reaction']=Diff_DF.Second_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(s)+'\")')\n",
    "Out_DF['Second Reaction Author']=Diff_DF.Second_Reaction.apply(lambda a: EventMap_Dict[a])\n",
    "Out_DF.drop('First_Reaction',inplace=True,axis=1)\n",
    "Out_DF.drop('Second_Reaction',inplace=True,axis=1)\n",
    "\n",
    "# write to file \n",
    "## path\n",
    "notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "path=notebook_path.rsplit('/',1)\n",
    "path=path[0]+'/Reaction_Connectivity/Pathways/'\n",
    "Out_DF.to_csv(path+'AllReactome'+'_MissingConnections.csv',index=False,header=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pathway connectivity analysis - author centric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import cypher\n",
    "import os\n",
    "\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\" #database connection\n",
    "\n",
    "#run query for pathway name-ID map\n",
    "PathwayName_ID_CQ=\"\"\"\n",
    "MATCH (pa:TopLevelPathway{speciesName:\"Homo sapiens\"})\n",
    "RETURN pa.displayName AS PathwayNames, pa.stId AS PathwayStIds\n",
    "\"\"\" \n",
    "PathwayMap_DF=cypher.run(PathwayName_ID_CQ,conn=Con).get_dataframe()\n",
    "PathwayMap_Dict=PathwayMap_DF.set_index('PathwayNames')['PathwayStIds'].to_dict()\n",
    "\n",
    "\n",
    "#run query for event-author map\n",
    "Event_Author_CQ=\"\"\"\n",
    "MATCH(pa:TopLevelPathway)-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "OPTIONAL MATCH (ev)-[:authored]-(ie:InstanceEdit)-[:author]-(au:Person)\n",
    "RETURN DISTINCT ev.stId AS EventID, au.displayName AS Author\n",
    "\"\"\"\n",
    "EventMap_DF=cypher.run(Event_Author_CQ,conn=Con).get_dataframe()\n",
    "EventMap_Dict=EventMap_DF.set_index('EventID')['Author'].to_dict()\n",
    "\n",
    "PathwaysOutputDF_List=[]\n",
    "\n",
    "#Analyze missing connections for each top-level pathway\n",
    "for i in sorted(PathwayMap_Dict):\n",
    "#     # skip big pathways to avoid memory issues\n",
    "#     if i in ['Immune System','Signal Transduction']:\n",
    "#         continue\n",
    "    \n",
    "    print(i)\n",
    "    \n",
    "    #assign pathway ID\n",
    "    Pathway_Name=i\n",
    "    Pathway_stId='\"'+PathwayMap_Dict[i]+'\"'\n",
    "    \n",
    "    #run query to get precedingEvents connections\n",
    "    Preced_CQ=\"\"\"\n",
    "    MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    MATCH(ev)-[:precedingEvent]->(pe:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    RETURN DISTINCT pe.stId AS First_Reaction, ev.stId AS Second_Reaction, pa.displayName AS Pathway\n",
    "    \"\"\"\n",
    "    Preced_DF=cypher.run(Preced_CQ,conn=Con).get_dataframe()\n",
    "    \n",
    "    if i=='Metabolism':\n",
    "        #query for shared entity connections in Metabolism factoring in simple entities\n",
    "        Shared_CQ=\"\"\"\n",
    "        MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "        MATCH(ro)-[:input|output|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate]->(po1:PhysicalEntity)-[:referenceEntity]->(ref0:ReferenceMolecule)\n",
    "        WHERE NOT ANY(x IN ref0.identifier WHERE x IN [\"15377\",\"16761\",\"15422\",\"16027\",\"15996\",\"17552\",\"17659\",\"15713\",\"15378\",\"16234\",\"18367\",\"29888\",\"13392\",\"18009\",\"16908\",\"16238\",\"17877\",\"15346\",\"15379\",\"16526\",\"15846\",\"17345\",\"16174\",\"17713\",\"16284\",\"28862\",\"16192\",\"16497\",\"17625\",\"15713\",\"16695\",\"18009\"])\n",
    "        WITH pa, ro, po1\n",
    "        MATCH(po1)<-[:output]-(ro)\n",
    "        WHERE NOT (po1.stId=\"R-HSA-113595\")\n",
    "        WITH pa, ro, po1\n",
    "        MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity]-(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)\n",
    "        WITH pa, ro, ri1\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction, pa.displayName AS Pathway\n",
    "        UNION MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2:PhysicalEntity{schemaClass:\"DefinedSet\"})-[:output|physicalEntity|hasMember|hasCandidate]->(po3:PhysicalEntity)\n",
    "        WHERE NOT (po2.stId=\"R-HSA-113595\")\n",
    "        WITH pa, ro, po3\n",
    "        MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate]-(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)\n",
    "        WITH pa, ro, ri2\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction, pa.displayName AS Pathway\n",
    "        \"\"\"\n",
    "    else:\n",
    "        #query for shared entity connections\n",
    "        Shared_CQ=\"\"\"\n",
    "        MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc1:PhysicalEntity)\n",
    "        WITH  pa, ro, COLLECT(DISTINCT dc1.schemaClass) AS alldc1\n",
    "        WHERE NOT (\"ChemicalDrug\" IN (alldc1) OR \"ProteinDrug\" IN (alldc1))\n",
    "        WITH pa, ro\n",
    "        MATCH(ro)-[:output]->(po1:PhysicalEntity)\n",
    "        WHERE NOT (po1.schemaClass=\"SimpleEntity\" OR po1.stId=\"R-HSA-113595\")\n",
    "        WITH pa, ro, po1\n",
    "        MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|hasMember|hasCandidate*]->(dc2:PhysicalEntity)\n",
    "        WITH  pa, ro, po1, ri1, COLLECT(DISTINCT dc2.schemaClass) AS alldc2\n",
    "        WHERE NOT (\"ChemicalDrug\" IN (alldc2) OR \"ProteinDrug\" IN (alldc2))\n",
    "        WITH pa, ro, po1, ri1\n",
    "        MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "        WITH pa, ro, ri1\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri1.stId AS Second_Reaction, pa.displayName AS Pathway\n",
    "        ORDER BY ro.stId\n",
    "        UNION MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2:PhysicalEntity{schemaClass:\"DefinedSet\"})-[:output|physicalEntity|hasMember|hasCandidate*]->(po3:PhysicalEntity)\n",
    "        WHERE NOT (po2.stId=\"R-HSA-113595\")\n",
    "        WITH pa, ro, po3\n",
    "        MATCH(po3)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)\n",
    "        WITH pa, ro, ri2\n",
    "        RETURN DISTINCT ro.stId AS First_Reaction, ri2.stId AS Second_Reaction, pa.displayName AS Pathway\n",
    "        ORDER BY ro.stId\n",
    "        \"\"\"\n",
    "        \n",
    "    #run query to get shared entity connections\n",
    "    Shared_DF=cypher.run(Shared_CQ,conn=Con).get_dataframe()\n",
    "    \n",
    "    # find missing reaction connections \n",
    "    Diff_DF=Shared_DF[~Shared_DF.apply(tuple,1).isin(Preced_DF.apply(tuple,1))]\n",
    "    print('# missing connections = '+str(len(Diff_DF)))\n",
    "\n",
    "    # generate hyperlinks\n",
    "    Out_DF=Diff_DF\n",
    "    Out_DF['First Reaction']=Diff_DF.First_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(s)+'\")')\n",
    "    Out_DF['Second Reaction']=Diff_DF.Second_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(s)+'\")')\n",
    "    Out_DF['Second Reaction Author']=Diff_DF.Second_Reaction.apply(lambda a: EventMap_Dict[a])\n",
    "    Out_DF.drop('First_Reaction',inplace=True,axis=1)\n",
    "    Out_DF.drop('Second_Reaction',inplace=True,axis=1)\n",
    "    \n",
    "    #append list\n",
    "    PathwaysOutputDF_List.append(Out_DF)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#concatenate pathways output\n",
    "PathwaysDF_Concat=pd.concat(PathwaysOutputDF_List)\n",
    "PathwaysDF_Concat.fillna('None',inplace=True)\n",
    "\n",
    "PathwaysDF_Concat[PathwaysDF_Concat['Second Reaction Author']=='None']\n",
    "\n",
    "#create author-specific table\n",
    "for a in PathwaysDF_Concat['Second Reaction Author'].unique():\n",
    "    if a in PathwaysDF_Concat['Second Reaction Author'].unique():\n",
    "        Author_DF=PathwaysDF_Concat[PathwaysDF_Concat['Second Reaction Author']==a]\n",
    "        # write to file \n",
    "        ## path\n",
    "        notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "        path=notebook_path.rsplit('/',1)\n",
    "        path=path[0]+'/Reaction_Connectivity/Authors/'\n",
    "        Author_DF.to_csv(path+a+'_Missing_precedingEvents.csv',index=False,header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Adding precedingEvent relation in Reactome curator project "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def precedingEvent_Adder(XML_Root, first_stID, first_class, second_stID):\n",
    "    #search for second reaction\n",
    "    c=0\n",
    "    for node in XML_Root.iter('instance'):\n",
    "        if node.attrib['DB_ID']==second_stID:\n",
    "            c=1\n",
    "            #print(node.attrib['displayName'])\n",
    "            #search for precedingEvent in second reaction\n",
    "            x=0\n",
    "            for subnode in node.iter('attribute'):\n",
    "                if subnode.attrib['name']=='precedingEvent':\n",
    "                    #print(subnode.attrib['referTo'])\n",
    "                    #check if second reaction already has the first reaction as a precedingEvent\n",
    "                    if subnode.attrib['referTo']==first_stID:\n",
    "                        x=1\n",
    "            #do nothing if first reaction is already a precedingEvent\n",
    "            if x==1: \n",
    "                #print(\"Already Exists!\")\n",
    "                break\n",
    "            #else add the first reaction to preceedingEvent attribute\n",
    "            else:\n",
    "                #print(\"New precedingEvent added!\")\n",
    "                tag=ET.SubElement(node,'attribute')\n",
    "                tag.attrib['name']='precedingEvent'\n",
    "                tag.attrib['class']=first_class\n",
    "                tag.attrib['referTo']=first_stID\n",
    "                node.attrib['isDirty']=\"true\"\n",
    "    if c==0:\n",
    "        print(second_stID,\"not found in project!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "#path\n",
    "notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "path=notebook_path.rsplit('/',1)\n",
    "path=path[0]+'/Reaction_Connectivity/'\n",
    "InputProject='Complete_ImmuneSystem_Input.rtpj'\n",
    "MyTree=ET.parse(path+InputProject)\n",
    "MyRoot=MyTree.getroot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#viewing the xml\n",
    "from xml.dom import minidom\n",
    "\n",
    "dom=minidom.parseString(ET.tostring(MyRoot))\n",
    "#print(dom.toprettyxml())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cypher\n",
    "\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\" #database connection\n",
    "\n",
    "#run query for pathway name-ID map\n",
    "ReaClass_ID_CQ=\"\"\"\n",
    "MATCH (ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "RETURN DISTINCT ev.stId AS EventID, ev.schemaClass AS Class\n",
    "\"\"\" \n",
    "ReaClassMap_DF=cypher.run(ReaClass_ID_CQ,conn=Con).get_dataframe()\n",
    "ReaClass_Dict=ReaClassMap_DF.set_index('EventID')['Class'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gspread\n",
    "from oauth2client.service_account import ServiceAccountCredentials\n",
    "\n",
    "# path\n",
    "notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "path=notebook_path.rsplit('/',1)\n",
    "path=path[0]+'/Reaction_Connectivity/'\n",
    "\n",
    "# define the scope\n",
    "scope = ['https://spreadsheets.google.com/feeds','https://www.googleapis.com/auth/drive']\n",
    "\n",
    "# add credentials to the account\n",
    "creds = ServiceAccountCredentials.from_json_keyfile_name(path+'Reactome Connectivity-f994dc05dc97.json', scope)\n",
    "\n",
    "# authorize the clientsheet \n",
    "client = gspread.authorize(creds)\n",
    "\n",
    "# get the instance of the Spreadsheet\n",
    "sheet = client.open('ImmuneSystem_Missing_precedingEvents')\n",
    "\n",
    "# get the first sheet of the Spreadsheet\n",
    "sheet_instance = sheet.get_worksheet(0)\n",
    "\n",
    "# get the total number of columns\n",
    "#sheet_instance.col_count\n",
    "\n",
    "\n",
    "# get the value at the specific cell\n",
    "#sheet_instance.cell(col=3,row=2)\n",
    "\n",
    "# get all the records of the data\n",
    "records_data = sheet_instance.get_all_records()\n",
    "\n",
    "# view the data\n",
    "#records_data\n",
    "\n",
    "# convert dict to dataframe\n",
    "records_df = pd.DataFrame.from_dict(records_data)\n",
    "\n",
    "# view top records\n",
    "#records_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test set\n",
    "#extracting approved instances\n",
    "Approved_Records=records_df[records_df['Approved/Rejected']=='Approved']\n",
    "Max_Freq_PrEv=Approved_Records['Event'].value_counts().index[0]\n",
    "Approved_Records=Approved_Records[Approved_Records['Event']==Max_Freq_PrEv]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Approved_Records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#extracting approved instances\n",
    "#Approved_Records=records_df[records_df['Approved/Rejected']=='Approved']\n",
    "\n",
    "#run analysis for all approved\n",
    "for index, row in Approved_Records.iterrows():\n",
    "    #print(row['precedingEvent'],row['Event'])\n",
    "    precedingEvent_Adder(MyRoot,row['precedingEvent'].split('R-HSA-')[1],ReaClass_Dict[row['precedingEvent']],row['Event'].split('R-HSA-')[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from xml.dom import minidom\n",
    "\n",
    "xmlstr = minidom.parseString(ET.tostring(MyRoot,'utf-8')).toprettyxml(indent=\" \",newl='')\n",
    "with open(path+\"Complete_ImmuneSystem_Output.rtpj\", \"w\") as f:\n",
    "    f.write(xmlstr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(Approved_Records)\n",
    "len(set(list(Approved_Records['Event'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Integrative Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following operations are performed in this section\n",
    "1. run cypher queries to identify connections between reactions within top-level pathways and within top-level pathways\n",
    "2. apply filters to remove connections based no simple entities and sets\n",
    "3. apply filters to remove drug-related events\n",
    "4. apply filters to remove cyclic events\n",
    "5. find connections that are not already curated\n",
    "6. create hyperlinks to corresponding reactions in Reactome website\n",
    "7. run cypher queries to get profile of reactions such as inputs, outputs, catalysts and regulators\n",
    "8. create Boolean relations for missing connections "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27 rows affected.\n",
      "14371 rows affected.\n",
      "13345 rows affected.\n",
      "20232 rows affected.\n",
      "1264 rows affected.\n",
      "Autophagy\n",
      "131 rows affected.\n",
      "169 rows affected.\n",
      "No filters applied:  (169, 5)\n",
      "After removing sets and simple entities:  (166, 5)\n",
      "After removing drug events:  (166, 5)\n",
      "After removing exclusion molecules:  (166, 5)\n",
      "After removing cyclic events:  (161, 5)\n",
      "After removing curated events:  (14, 6)\n",
      "110 rows affected.\n",
      "Pathway Name:  Autophagy  Progress:  1 / 27  Network size:  161\n",
      "Cell Cycle\n",
      "467 rows affected.\n",
      "948 rows affected.\n",
      "No filters applied:  (948, 5)\n",
      "After removing sets and simple entities:  (947, 5)\n",
      "After removing drug events:  (947, 5)\n",
      "After removing exclusion molecules:  (947, 5)\n",
      "After removing cyclic events:  (832, 5)\n",
      "After removing curated events:  (494, 6)\n",
      "461 rows affected.\n",
      "Pathway Name:  Cell Cycle  Progress:  2 / 27  Network size:  832\n",
      "Cell-Cell communication\n",
      "25 rows affected.\n",
      "24 rows affected.\n",
      "No filters applied:  (24, 5)\n",
      "After removing sets and simple entities:  (24, 5)\n",
      "After removing drug events:  (24, 5)\n",
      "After removing exclusion molecules:  (24, 5)\n",
      "After removing cyclic events:  (24, 5)\n",
      "After removing curated events:  (1, 6)\n",
      "60 rows affected.\n",
      "Pathway Name:  Cell-Cell communication  Progress:  3 / 27  Network size:  24\n",
      "Cellular responses to external stimuli\n",
      "244 rows affected.\n",
      "395 rows affected.\n",
      "No filters applied:  (395, 5)\n",
      "After removing sets and simple entities:  (328, 5)\n",
      "After removing drug events:  (326, 5)\n",
      "After removing exclusion molecules:  (326, 5)\n",
      "After removing cyclic events:  (277, 5)\n",
      "After removing curated events:  (127, 6)\n",
      "280 rows affected.\n",
      "Pathway Name:  Cellular responses to external stimuli  Progress:  4 / 27  Network size:  277\n",
      "Chromatin organization\n",
      "8 rows affected.\n",
      "374 rows affected.\n",
      "No filters applied:  (374, 5)\n",
      "After removing sets and simple entities:  (374, 5)\n",
      "After removing drug events:  (373, 5)\n",
      "After removing exclusion molecules:  (373, 5)\n",
      "After removing cyclic events:  (249, 5)\n",
      "After removing curated events:  (242, 6)\n",
      "85 rows affected.\n",
      "Pathway Name:  Chromatin organization  Progress:  5 / 27  Network size:  249\n",
      "Circadian Clock\n",
      "69 rows affected.\n",
      "64 rows affected.\n",
      "No filters applied:  (64, 5)\n",
      "After removing sets and simple entities:  (64, 5)\n",
      "After removing drug events:  (64, 5)\n",
      "After removing exclusion molecules:  (64, 5)\n",
      "After removing cyclic events:  (58, 5)\n",
      "After removing curated events:  (4, 6)\n",
      "81 rows affected.\n",
      "Pathway Name:  Circadian Clock  Progress:  6 / 27  Network size:  58\n",
      "DNA Repair\n",
      "322 rows affected.\n",
      "1430 rows affected.\n",
      "No filters applied:  (1430, 5)\n",
      "After removing sets and simple entities:  (1308, 5)\n",
      "After removing drug events:  (1308, 5)\n",
      "After removing exclusion molecules:  (1308, 5)\n",
      "After removing cyclic events:  (1234, 5)\n",
      "After removing curated events:  (949, 6)\n",
      "331 rows affected.\n",
      "Pathway Name:  DNA Repair  Progress:  7 / 27  Network size:  1234\n",
      "DNA Replication\n",
      "45 rows affected.\n",
      "45 rows affected.\n",
      "No filters applied:  (45, 5)\n",
      "After removing sets and simple entities:  (45, 5)\n",
      "After removing drug events:  (45, 5)\n",
      "After removing exclusion molecules:  (45, 5)\n",
      "After removing cyclic events:  (43, 5)\n",
      "After removing curated events:  (7, 6)\n",
      "47 rows affected.\n",
      "Pathway Name:  DNA Replication  Progress:  8 / 27  Network size:  43\n",
      "Developmental Biology\n",
      "460 rows affected.\n",
      "702 rows affected.\n",
      "No filters applied:  (702, 5)\n",
      "After removing sets and simple entities:  (641, 5)\n",
      "After removing drug events:  (637, 5)\n",
      "After removing exclusion molecules:  (637, 5)\n",
      "After removing cyclic events:  (591, 5)\n",
      "After removing curated events:  (202, 6)\n",
      "602 rows affected.\n",
      "Pathway Name:  Developmental Biology  Progress:  9 / 27  Network size:  591\n",
      "Digestion and absorption\n",
      "32 rows affected.\n",
      "29 rows affected.\n",
      "No filters applied:  (29, 5)\n",
      "After removing sets and simple entities:  (29, 5)\n",
      "After removing drug events:  (29, 5)\n",
      "After removing exclusion molecules:  (29, 5)\n",
      "After removing cyclic events:  (25, 5)\n",
      "After removing curated events:  (3, 6)\n",
      "30 rows affected.\n",
      "Pathway Name:  Digestion and absorption  Progress:  10 / 27  Network size:  25\n",
      "Disease\n",
      "954 rows affected.\n",
      "4209 rows affected.\n",
      "No filters applied:  (4209, 5)\n",
      "After removing sets and simple entities:  (2693, 5)\n",
      "After removing drug events:  (2308, 5)\n",
      "After removing exclusion molecules:  (2308, 5)\n",
      "After removing cyclic events:  (2194, 5)\n",
      "After removing curated events:  (1101, 6)\n",
      "1575 rows affected.\n",
      "Pathway Name:  Disease  Progress:  11 / 27  Network size:  2194\n",
      "Extracellular matrix organization\n",
      "297 rows affected.\n",
      "1286 rows affected.\n",
      "No filters applied:  (1286, 5)\n",
      "After removing sets and simple entities:  (1284, 5)\n",
      "After removing drug events:  (1282, 5)\n",
      "After removing exclusion molecules:  (1282, 5)\n",
      "After removing cyclic events:  (1274, 5)\n",
      "After removing curated events:  (246, 6)\n",
      "320 rows affected.\n",
      "Pathway Name:  Extracellular matrix organization  Progress:  12 / 27  Network size:  1274\n",
      "Gene expression (Transcription)\n",
      "1098 rows affected.\n",
      "2775 rows affected.\n",
      "No filters applied:  (2775, 5)\n",
      "After removing sets and simple entities:  (2346, 5)\n",
      "After removing drug events:  (2346, 5)\n",
      "After removing exclusion molecules:  (2346, 5)\n",
      "After removing cyclic events:  (2176, 5)\n",
      "After removing curated events:  (1288, 6)\n",
      "1059 rows affected.\n",
      "Pathway Name:  Gene expression (Transcription)  Progress:  13 / 27  Network size:  2176\n",
      "Hemostasis\n",
      "269 rows affected.\n",
      "481 rows affected.\n",
      "No filters applied:  (481, 5)\n",
      "After removing sets and simple entities:  (372, 5)\n",
      "After removing drug events:  (316, 5)\n",
      "After removing exclusion molecules:  (316, 5)\n",
      "After removing cyclic events:  (286, 5)\n",
      "After removing curated events:  (114, 6)\n",
      "347 rows affected.\n",
      "Pathway Name:  Hemostasis  Progress:  14 / 27  Network size:  286\n",
      "Immune System\n",
      "1438 rows affected.\n",
      "4199 rows affected.\n",
      "No filters applied:  (4199, 5)\n",
      "After removing sets and simple entities:  (3141, 5)\n",
      "After removing drug events:  (3031, 5)\n",
      "After removing exclusion molecules:  (3031, 5)\n",
      "After removing cyclic events:  (2718, 5)\n",
      "After removing curated events:  (1574, 6)\n",
      "1685 rows affected.\n",
      "Pathway Name:  Immune System  Progress:  15 / 27  Network size:  2718\n",
      "Metabolism\n",
      "2391 rows affected.\n",
      "71229 rows affected.\n",
      "No filters applied:  (71229, 5)\n",
      "After removing sets and simple entities:  (71197, 5)\n",
      "After removing drug events:  (70066, 5)\n",
      "After removing exclusion molecules:  (14621, 5)\n",
      "After removing cyclic events:  (10986, 5)\n",
      "After removing curated events:  (8508, 6)\n",
      "2327 rows affected.\n",
      "Pathway Name:  Metabolism  Progress:  16 / 27  Network size:  10986\n",
      "Metabolism of RNA\n",
      "181 rows affected.\n",
      "262 rows affected.\n",
      "No filters applied:  (262, 5)\n",
      "After removing sets and simple entities:  (257, 5)\n",
      "After removing drug events:  (257, 5)\n",
      "After removing exclusion molecules:  (257, 5)\n",
      "After removing cyclic events:  (228, 5)\n",
      "After removing curated events:  (124, 6)\n",
      "236 rows affected.\n",
      "Pathway Name:  Metabolism of RNA  Progress:  17 / 27  Network size:  228\n",
      "Metabolism of proteins\n",
      "733 rows affected.\n",
      "2685 rows affected.\n",
      "No filters applied:  (2685, 5)\n",
      "After removing sets and simple entities:  (2217, 5)\n",
      "After removing drug events:  (2189, 5)\n",
      "After removing exclusion molecules:  (2189, 5)\n",
      "After removing cyclic events:  (1814, 5)\n",
      "After removing curated events:  (1183, 6)\n",
      "923 rows affected.\n",
      "Pathway Name:  Metabolism of proteins  Progress:  18 / 27  Network size:  1814\n",
      "Muscle contraction\n",
      "22 rows affected.\n",
      "111 rows affected.\n",
      "No filters applied:  (111, 5)\n",
      "After removing sets and simple entities:  (21, 5)\n",
      "After removing drug events:  (19, 5)\n",
      "After removing exclusion molecules:  (19, 5)\n",
      "After removing cyclic events:  (19, 5)\n",
      "After removing curated events:  (9, 6)\n",
      "59 rows affected.\n",
      "Pathway Name:  Muscle contraction  Progress:  19 / 27  Network size:  19\n",
      "Neuronal System\n",
      "140 rows affected.\n",
      "602 rows affected.\n",
      "No filters applied:  (602, 5)\n",
      "After removing sets and simple entities:  (275, 5)\n",
      "After removing drug events:  (272, 5)\n",
      "After removing exclusion molecules:  (272, 5)\n",
      "After removing cyclic events:  (212, 5)\n",
      "After removing curated events:  (129, 6)\n",
      "217 rows affected.\n",
      "Pathway Name:  Neuronal System  Progress:  20 / 27  Network size:  212\n",
      "Organelle biogenesis and maintenance\n",
      "76 rows affected.\n",
      "90 rows affected.\n",
      "No filters applied:  (90, 5)\n",
      "After removing sets and simple entities:  (88, 5)\n",
      "After removing drug events:  (88, 5)\n",
      "After removing exclusion molecules:  (88, 5)\n",
      "After removing cyclic events:  (84, 5)\n",
      "After removing curated events:  (26, 6)\n",
      "95 rows affected.\n",
      "Pathway Name:  Organelle biogenesis and maintenance  Progress:  21 / 27  Network size:  84\n",
      "Programmed Cell Death\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "233 rows affected.\n",
      "251 rows affected.\n",
      "No filters applied:  (251, 5)\n",
      "After removing sets and simple entities:  (251, 5)\n",
      "After removing drug events:  (243, 5)\n",
      "After removing exclusion molecules:  (243, 5)\n",
      "After removing cyclic events:  (223, 5)\n",
      "After removing curated events:  (71, 6)\n",
      "186 rows affected.\n",
      "Pathway Name:  Programmed Cell Death  Progress:  22 / 27  Network size:  223\n",
      "Protein localization\n",
      "49 rows affected.\n",
      "102 rows affected.\n",
      "No filters applied:  (102, 5)\n",
      "After removing sets and simple entities:  (102, 5)\n",
      "After removing drug events:  (102, 5)\n",
      "After removing exclusion molecules:  (102, 5)\n",
      "After removing cyclic events:  (83, 5)\n",
      "After removing curated events:  (32, 6)\n",
      "53 rows affected.\n",
      "Pathway Name:  Protein localization  Progress:  23 / 27  Network size:  83\n",
      "Reproduction\n",
      "22 rows affected.\n",
      "23 rows affected.\n",
      "No filters applied:  (23, 5)\n",
      "After removing sets and simple entities:  (23, 5)\n",
      "After removing drug events:  (23, 5)\n",
      "After removing exclusion molecules:  (23, 5)\n",
      "After removing cyclic events:  (18, 5)\n",
      "After removing curated events:  (6, 6)\n",
      "25 rows affected.\n",
      "Pathway Name:  Reproduction  Progress:  24 / 27  Network size:  18\n",
      "Signal Transduction\n",
      "2445 rows affected.\n",
      "7258 rows affected.\n",
      "No filters applied:  (7258, 5)\n",
      "After removing sets and simple entities:  (5300, 5)\n",
      "After removing drug events:  (5132, 5)\n",
      "After removing exclusion molecules:  (5132, 5)\n",
      "After removing cyclic events:  (4739, 5)\n",
      "After removing curated events:  (2834, 6)\n",
      "2522 rows affected.\n",
      "Pathway Name:  Signal Transduction  Progress:  25 / 27  Network size:  4739\n",
      "Transport of small molecules\n",
      "165 rows affected.\n",
      "4324 rows affected.\n",
      "No filters applied:  (4324, 5)\n",
      "After removing sets and simple entities:  (191, 5)\n",
      "After removing drug events:  (178, 5)\n",
      "After removing exclusion molecules:  (178, 5)\n",
      "After removing cyclic events:  (142, 5)\n",
      "After removing curated events:  (70, 6)\n",
      "458 rows affected.\n",
      "Pathway Name:  Transport of small molecules  Progress:  26 / 27  Network size:  142\n",
      "Vesicle-mediated transport\n",
      "162 rows affected.\n",
      "486 rows affected.\n",
      "No filters applied:  (486, 5)\n",
      "After removing sets and simple entities:  (481, 5)\n",
      "After removing drug events:  (478, 5)\n",
      "After removing exclusion molecules:  (478, 5)\n",
      "After removing cyclic events:  (369, 5)\n",
      "After removing curated events:  (254, 6)\n",
      "260 rows affected.\n",
      "Pathway Name:  Vesicle-mediated transport  Progress:  27 / 27  Network size:  369\n"
     ]
    }
   ],
   "source": [
    "# top-level pathways\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cypher\n",
    "import os\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\" #database connection\n",
    "\n",
    "#run query for pathway name-ID map\n",
    "PathwayName_ID_CQ=\"\"\"\n",
    "MATCH (pa:TopLevelPathway{speciesName:\"Homo sapiens\"})\n",
    "RETURN pa.displayName AS PathwayNames, pa.stId AS PathwayStIds\n",
    "\"\"\" \n",
    "PathwayMap_DF=cypher.run(PathwayName_ID_CQ,conn=Con).get_dataframe()\n",
    "PathwayMap_Dict=PathwayMap_DF.set_index('PathwayNames')['PathwayStIds'].to_dict()\n",
    "\n",
    "#run query for event-author map\n",
    "Event_Author_CQ=\"\"\"\n",
    "MATCH(pa:TopLevelPathway)-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "OPTIONAL MATCH (ev)-[:authored]-(ie:InstanceEdit)-[:author]-(au:Person)\n",
    "RETURN DISTINCT ev.stId AS EventID, au.displayName AS Author\n",
    "\"\"\"\n",
    "EventMap_DF=cypher.run(Event_Author_CQ,conn=Con).get_dataframe()\n",
    "EventMap_Dict=EventMap_DF.set_index('EventID')['Author'].to_dict()\n",
    "\n",
    "#run query for reaction ID-name map\n",
    "ReactionIDName_CQ=\"\"\"\n",
    "MATCH (re:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "RETURN DISTINCT re.stId AS ReactionID, re.displayName AS ReactionName\n",
    "\"\"\"\n",
    "ReacMap_DF=cypher.run(ReactionIDName_CQ,conn=Con).get_dataframe()\n",
    "ReacMap_Dict=ReacMap_DF.set_index('ReactionID')['ReactionName'].to_dict()\n",
    "\n",
    "#run query for reference ID:Reactome ID map\n",
    "IDmap_CQ=\"\"\"\n",
    "MATCH(pa:TopLevelPathway{speciesName:\"Homo sapiens\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|output|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]->(po:PhysicalEntity)-[:referenceEntity]->(ref:ReferenceMolecule)\n",
    "RETURN po.displayName AS Name,ref.identifier AS Ref_Id, po.stId AS React_Id\n",
    "\"\"\"\n",
    "IDmap_DF=cypher.run(IDmap_CQ,conn=Con).get_dataframe()\n",
    "IDmap_Dict=IDmap_DF.set_index('Ref_Id')['React_Id'].to_dict()\n",
    "\n",
    "#run query to get all drug-related reactions\n",
    "DrugReactions_CQ=\"\"\"\n",
    "MATCH(p:TopLevelPathway)-[:hasEvent*]->(r:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|output|catalystActivity|hasComponent|hasMember|hasCandidate|physicalEntity|regulatedBy|regulator*]->(d:PhysicalEntity)\n",
    "WHERE (d.schemaClass='ChemicalDrug' OR d.schemaClass='ProteinDrug')\n",
    "RETURN DISTINCT p.displayName AS PathwayName, r.stId AS ReactionID, r.displayName AS ReactionName, d.displayName AS DrugName, d.schemaClass AS DrugType\n",
    "\"\"\"\n",
    "DrugReactions_DF=cypher.run(DrugReactions_CQ,conn=Con).get_dataframe()\n",
    "\n",
    "\n",
    "## path\n",
    "notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "path=notebook_path.rsplit('/',1)\n",
    "path1=path[0]+'/Reaction_Connectivity/'\n",
    "path2=path[0]+'/Integrative_Analysis/'\n",
    "## read exclusion list\n",
    "wb=load_workbook(path1+'Exclusion_Molecules.xlsx',read_only=True)\n",
    "\n",
    "\n",
    "#Analyze missing connections for each top-level pathway\n",
    "m=0\n",
    "for i in sorted(PathwayMap_Dict):\n",
    "#for i in ['Hemostasis']:\n",
    "\n",
    "#     # skip big pathways to avoid memory issues\n",
    "#     if i in ['Immune System','Signal Transduction']:\n",
    "#         continue\n",
    "    \n",
    "    print(i)\n",
    "    \n",
    "    #assign pathway ID\n",
    "    Pathway_Name=i\n",
    "    Pathway_stId='\"'+PathwayMap_Dict[i]+'\"'\n",
    "    \n",
    "    #run query to get precedingEvents connections\n",
    "    Preced_CQ=\"\"\"\n",
    "    MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    MATCH(ev)-[:precedingEvent]->(pe:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    RETURN DISTINCT pe.stId AS First_Reaction, ev.stId AS Second_Reaction\n",
    "    \"\"\"\n",
    "    Preced_DF=cypher.run(Preced_CQ,conn=Con).get_dataframe()\n",
    "    \n",
    "    #run query to get reactions connected by shared entities\n",
    "    Shared_CQ=\"\"\"\n",
    "    ///query for non-set reactions\n",
    "    MATCH(pa1:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po1:PhysicalEntity)\n",
    "    WHERE NOT (po1.schemaClass=\"DefinedSet\" OR po1.schemaClass=\"CandidateSet\" OR po1.stId=\"R-HSA-113595\") //ignore Ub\n",
    "    WITH pa1, ro1, po1\n",
    "    MATCH(pa1)-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    WITH ro1, po1, ri1\n",
    "    MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "    WITH ro1, ri1, po1\n",
    "    RETURN DISTINCT ro1.stId AS First_Reaction, ri1.stId AS Second_Reaction, po1.schemaClass AS SharedEntityClass, po1.displayName AS SharedEntityName, po1.stId AS SharedEntityID\n",
    "    ORDER BY ro1.stId\n",
    "    //query for set connectors\n",
    "    UNION MATCH(pa2:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2a:PhysicalEntity)-[:hasMember|hasCandidate|physicalEntity*]->(po2b:PhysicalEntity)\n",
    "    WHERE (po2a.schemaClass=\"DefinedSet\" OR po2a.schemaClass=\"CandidateSet\") AND NOT (po2a.stId=\"R-HSA-113595\") //ignore Ub\n",
    "    WITH pa2, ro2, po2b\n",
    "    MATCH(pa2)-[:hasEvent*]->(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    MATCH(po2b)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2)\n",
    "    WITH ro2, ri2, po2b\n",
    "    RETURN DISTINCT ro2.stId AS First_Reaction, ri2.stId AS Second_Reaction, po2b.schemaClass AS SharedEntityClass, po2b.displayName AS SharedEntityName, po2b.stId AS SharedEntityID\n",
    "    ORDER BY ro2.stId\n",
    "    \"\"\"   \n",
    "    Shared_DF=cypher.run(Shared_CQ,conn=Con).get_dataframe()\n",
    "    #print('Shared_DF size: ',Shared_DF.shape)\n",
    "    \n",
    "# apply filters to exclude cyclic reactions\n",
    "\n",
    "    #read molecules to be excluded\n",
    "    SheetName=Pathway_Name\n",
    "    if SheetName in wb:\n",
    "        mol_df=pd.read_excel(path1+'Exclusion_Molecules.xlsx',sheet_name=Pathway_Name)\n",
    "        Exclude_List=list(map(str,list(mol_df['identifier'])))\n",
    "    else:\n",
    "        Exclude_List=[]\n",
    "        \n",
    "    print('No filters applied: ', Shared_DF.shape)\n",
    "    \n",
    "    #remove sets in shared entity column\n",
    "    df=Shared_DF[~Shared_DF['SharedEntityClass'].isin(['DefinedSet','CandidateSet'])]\n",
    "    if Pathway_Name not in ['Metabolism','Digestion and absorption']:\n",
    "        df=Shared_DF[~Shared_DF['SharedEntityClass'].isin(['SimpleEntity'])]\n",
    "    print('After removing sets and simple entities: ',df.shape)\n",
    "    \n",
    "    #remove drug-related events from First_Reaction and Second_Reaction\n",
    "    DrugReactions=set(DrugReactions_DF['ReactionID'])\n",
    "    df=df[~df['First_Reaction'].isin(DrugReactions)]\n",
    "    df=df[~df['Second_Reaction'].isin(DrugReactions)]\n",
    "    print('After removing drug events: ',df.shape)\n",
    "    \n",
    "    #remove exclusion molecules from shared entity column\n",
    "    Exclude_ID=set([IDmap_Dict[i] for i in Exclude_List if i in IDmap_Dict])\n",
    "    df=df[~df['SharedEntityID'].isin(Exclude_ID)]\n",
    "    print('After removing exclusion molecules: ',df.shape)\n",
    "    \n",
    "    #remove cyclic reactions\n",
    "    Index_List=[]\n",
    "    com1=df['First_Reaction']+'---'+df['Second_Reaction'] #creating reaction pair\n",
    "    com2=df['Second_Reaction']+'---'+df['First_Reaction'] #creating reverse reaction pair\n",
    "    #check if reaction pair equals reverse reaction pair\n",
    "    for ind,val in enumerate(com1):\n",
    "        if val in list(com2):\n",
    "            Index_List.append(ind)\n",
    "    df=df.drop(df.index[Index_List])\n",
    "    print('After removing cyclic events: ',df.shape)\n",
    "\n",
    "    #find missing reaction connections \n",
    "    df['Curated?']=df[['First_Reaction','Second_Reaction']].apply(tuple,1).isin(Preced_DF.apply(tuple,1))\n",
    "    print('After removing curated events: ',df[~df['Curated?']].shape)\n",
    "\n",
    "    # generate hyperlinks\n",
    "    df['First Reaction']=df.First_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(ReacMap_Dict[s])+'\")')\n",
    "    df['Second Reaction']=df.Second_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(ReacMap_Dict[s])+'\")')\n",
    "    df['Second Reaction Author']=df.Second_Reaction.apply(lambda a: EventMap_Dict[a])\n",
    "\n",
    "    #run query to generate nodes and edges for logic network model\n",
    "    Logic_Nodes_CQ=\"\"\"\n",
    "    MATCH(tpa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    WITH ev\n",
    "    OPTIONAL MATCH(ev)-[:input]->(pin:PhysicalEntity)\n",
    "    OPTIONAL MATCH (pin)-[:hasMember|hasCandidate*]->(pins:PhysicalEntity)\n",
    "    WITH ev,pin,pins\n",
    "    OPTIONAL MATCH(ev)-[:output]->(pou:PhysicalEntity)\n",
    "    OPTIONAL MATCH (pou)-[:hasMember|hasCandidate*]->(pous:PhysicalEntity)\n",
    "    WITH ev,pin,pins,pou,pous\n",
    "    OPTIONAL MATCH(ev)-[:catalystActivity|physicalEntity*]->(pca:PhysicalEntity)\n",
    "    OPTIONAL MATCH(pca)-[:hasMember|hasCandidate*]->(pcas:PhysicalEntity)\n",
    "    WITH ev,pin,pins,pou,pous,pca,pcas\n",
    "    OPTIONAL MATCH(ev)-[:regulatedBy|physicalEntity*]->(reg:Regulation)-[:regulator]->(pre:PhysicalEntity)\n",
    "    OPTIONAL MATCH(pre)-[:hasMember|hasCandidate*]->(pres:PhysicalEntity)\n",
    "    WITH ev, reg, \n",
    "    COLLECT(pin.stId)+COLLECT(pins.stId) AS i, COLLECT(pou.stId)+COLLECT(pous.stId) AS o, COLLECT(pca.stId)+COLLECT(pcas.stId) AS c, COLLECT(pre.stId)+COLLECT(pres.stId) AS r\n",
    "    RETURN DISTINCT ev.stId AS Reaction, i AS Input, o AS Output, c as Catalyst, r AS Regulator, reg.displayName AS Regulation\n",
    "    \"\"\"\n",
    "    Logic_Nodes_DF=cypher.run(Logic_Nodes_CQ,conn=Con).get_dataframe()\n",
    "\n",
    "    #data processing\n",
    "    ##creating logic rules and induction/supression value for edges\n",
    "    Output=[]\n",
    "    Input_List=[y for x in list(Logic_Nodes_DF['Input']) for y in x]\n",
    "    InputCounts={i:Input_List.count(i) for i in set(Input_List)}\n",
    "    for label,row in df.iterrows():\n",
    "        SharedEntity=row['SharedEntityID']\n",
    "        #label shared entity\n",
    "        if any(SharedEntity in x for x in list(Logic_Nodes_DF['Input'])):\n",
    "            SharedType='Input'\n",
    "        elif any(SharedEntity in x for x in list(Logic_Nodes_DF['Catalyst'])):\n",
    "            SharedType='Catalyst'\n",
    "        elif any(SharedEntity in x for x in list(Logic_Nodes_DF['Regulator'])):\n",
    "            RegulationType=[Logic_Nodes_DF.iloc[i]['Regulation'][:8] for i,v in Logic_Nodes_DF[['Regulator']].iterrows() if SharedEntity in v['Regulator']][0]\n",
    "            SharedType=RegulationType+' Regulator'\n",
    "        else:\n",
    "            print(\"Shared entity not found!\")\n",
    "        #assign Boolean logic and activation/suppression \n",
    "        if SharedType=='Input':\n",
    "            if InputCounts[SharedEntity]==1:\n",
    "                Logic='AND'\n",
    "                Value=1\n",
    "            elif InputCounts[SharedEntity]>1:\n",
    "                Logic='OR'\n",
    "                Value=0\n",
    "        elif SharedType!='Input':\n",
    "            Logic='AND'\n",
    "            if 'Positive' in SharedType:\n",
    "                Value=1\n",
    "            elif 'Negative' in SharedType:\n",
    "                Value=-1\n",
    "            elif 'Catalyst' in SharedType:\n",
    "                Value=1\n",
    "        Output.append([row['First Reaction'],row['Second Reaction'],row['Second Reaction Author'],SharedEntity,row['SharedEntityClass'],SharedType,Logic,Value,row['Curated?']])\n",
    "    Output_df=pd.DataFrame(Output,columns=['First_Reaction','Second_Reaction','Second_Reaction_Author','Shared_Entity','Shared_Entity_Class','Shared_Entity_Relation_to_Second_Reaction','Logic','Value','Is _It_Curated?'])\n",
    "    m+=1\n",
    "    print('Pathway Name: ',i,' Progress: ', m,'/',len(PathwayMap_Dict),' Network size: ',Output_df.shape[0])\n",
    "    #print('Network size',BoolEdges.shape)\n",
    "    #BoolEdges.head()\n",
    "\n",
    "    #write output to file\n",
    "    Output_df.to_csv(path2+Pathway_Name+'_IA.csv',index=False,header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13813 rows affected.\n",
      "14371 rows affected.\n",
      "13345 rows affected.\n",
      "20232 rows affected.\n",
      "1264 rows affected.\n",
      "11619 rows affected.\n",
      "103749 rows affected.\n",
      "No filters applied:  (103749, 5)\n",
      "After removing sets and simple entities:  (23689, 5)\n",
      "After removing intra-pathway connections:  (1031, 7)\n",
      "After removing drug events:  (1017, 7)\n",
      "After removing exclusion molecules:  (1017, 7)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-15-387d11f4903f>:128: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['First_Reaction_Pathway']=df['First_Reaction'].map(PathReacMap_Dict)\n",
      "<ipython-input-15-387d11f4903f>:129: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df['Second_Reaction_Pathway']=df['Second_Reaction'].map(PathReacMap_Dict)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After removing cyclic events:  (976, 7)\n",
      "After removing curated events:  (821, 8)\n",
      "13691 rows affected.\n"
     ]
    }
   ],
   "source": [
    "# inter-pathway connections\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cypher\n",
    "import os\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\" #database connection\n",
    "\n",
    "#run query for reaction ID-name map\n",
    "PathwayReaction_CQ=\"\"\"\n",
    "MATCH (tpa:TopLevelPathway{speciesName:\"Homo sapiens\"})-[:hasEvent*]->(re:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "RETURN DISTINCT tpa.displayName AS PathwayName, re.displayName AS ReactionName, re.stId AS ReactionID\n",
    "\"\"\"\n",
    "PathReacMap_DF=cypher.run(PathwayReaction_CQ,conn=Con).get_dataframe()\n",
    "PathReacMap_Dict=PathReacMap_DF.set_index('ReactionID')['PathwayName'].to_dict()\n",
    "\n",
    "#run query for event-author map\n",
    "Event_Author_CQ=\"\"\"\n",
    "MATCH(pa:TopLevelPathway)-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "OPTIONAL MATCH (ev)-[:authored]-(ie:InstanceEdit)-[:author]-(au:Person)\n",
    "RETURN DISTINCT ev.stId AS EventID, au.displayName AS Author\n",
    "\"\"\"\n",
    "EventMap_DF=cypher.run(Event_Author_CQ,conn=Con).get_dataframe()\n",
    "EventMap_Dict=EventMap_DF.set_index('EventID')['Author'].to_dict()\n",
    "\n",
    "#run query for reaction ID-name map\n",
    "ReactionIDName_CQ=\"\"\"\n",
    "MATCH (re:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "RETURN DISTINCT re.stId AS ReactionID, re.displayName AS ReactionName\n",
    "\"\"\"\n",
    "ReacMap_DF=cypher.run(ReactionIDName_CQ,conn=Con).get_dataframe()\n",
    "ReacMap_Dict=ReacMap_DF.set_index('ReactionID')['ReactionName'].to_dict()\n",
    "\n",
    "#run query for reference ID:Reactome ID map\n",
    "IDmap_CQ=\"\"\"\n",
    "MATCH(pa:TopLevelPathway{speciesName:\"Homo sapiens\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|output|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]->(po:PhysicalEntity)-[:referenceEntity]->(ref:ReferenceMolecule)\n",
    "RETURN po.displayName AS Name,ref.identifier AS Ref_Id, po.stId AS React_Id\n",
    "\"\"\"\n",
    "IDmap_DF=cypher.run(IDmap_CQ,conn=Con).get_dataframe()\n",
    "IDmap_Dict=IDmap_DF.set_index('Ref_Id')['React_Id'].to_dict()\n",
    "\n",
    "#run query to get all drug-related reactions\n",
    "DrugReactions_CQ=\"\"\"\n",
    "MATCH(p:TopLevelPathway)-[:hasEvent*]->(r:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|output|catalystActivity|hasComponent|hasMember|hasCandidate|physicalEntity|regulatedBy|regulator*]->(d:PhysicalEntity)\n",
    "WHERE (d.schemaClass='ChemicalDrug' OR d.schemaClass='ProteinDrug')\n",
    "RETURN DISTINCT p.displayName AS PathwayName, r.stId AS ReactionID, r.displayName AS ReactionName, d.displayName AS DrugName, d.schemaClass AS DrugType\n",
    "\"\"\"\n",
    "DrugReactions_DF=cypher.run(DrugReactions_CQ,conn=Con).get_dataframe()\n",
    "\n",
    "\n",
    "## path\n",
    "notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "path=notebook_path.rsplit('/',1)\n",
    "path1=path[0]+'/Reaction_Connectivity/'\n",
    "path2=path[0]+'/Integrative_Analysis/'\n",
    "## read exclusion list\n",
    "wb=load_workbook(path1+'Exclusion_Molecules.xlsx',read_only=True)\n",
    "\n",
    "\n",
    "#Analyze missing connections for all of Reactome\n",
    "\n",
    "#assign pathway ID\n",
    "Pathway_Name='Inter-pathway_connections'\n",
    "#Pathway_stId='\"'+PathwayMap_Dict[i]+'\"'\n",
    "\n",
    "#run query to get precedingEvents connections\n",
    "Preced_CQ=\"\"\"\n",
    "MATCH(pa:TopLevelPathway)-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "MATCH(ev)-[:precedingEvent]->(pe:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "RETURN DISTINCT pe.stId AS First_Reaction, ev.stId AS Second_Reaction\n",
    "\"\"\"\n",
    "Preced_DF=cypher.run(Preced_CQ,conn=Con).get_dataframe()\n",
    "\n",
    "#run query to get reactions connected by shared entities\n",
    "Shared_CQ=\"\"\"\n",
    "///query for non-set reactions\n",
    "MATCH(pa1:TopLevelPathway)-[:hasEvent*]->(ro1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po1:PhysicalEntity)\n",
    "WHERE NOT (po1.schemaClass=\"DefinedSet\" OR po1.schemaClass=\"CandidateSet\" OR po1.stId=\"R-HSA-113595\") //ignore Ub\n",
    "WITH pa1, ro1, po1\n",
    "MATCH(pa1)-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "WITH ro1, po1, ri1\n",
    "MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "WITH ro1, ri1, po1\n",
    "RETURN DISTINCT ro1.stId AS First_Reaction, ri1.stId AS Second_Reaction, po1.schemaClass AS SharedEntityClass, po1.displayName AS SharedEntityName, po1.stId AS SharedEntityID\n",
    "ORDER BY ro1.stId\n",
    "//query for set connectors\n",
    "UNION MATCH(pa2:TopLevelPathway)-[:hasEvent*]->(ro2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2a:PhysicalEntity)-[:hasMember|hasCandidate|physicalEntity*]->(po2b:PhysicalEntity)\n",
    "WHERE (po2a.schemaClass=\"DefinedSet\" OR po2a.schemaClass=\"CandidateSet\") AND NOT (po2a.stId=\"R-HSA-113595\") //ignore Ub\n",
    "WITH pa2, ro2, po2b\n",
    "MATCH(pa2)-[:hasEvent*]->(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "MATCH(po2b)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2)\n",
    "WITH ro2, ri2, po2b\n",
    "RETURN DISTINCT ro2.stId AS First_Reaction, ri2.stId AS Second_Reaction, po2b.schemaClass AS SharedEntityClass, po2b.displayName AS SharedEntityName, po2b.stId AS SharedEntityID\n",
    "ORDER BY ro2.stId\n",
    "\"\"\"   \n",
    "Shared_DF=cypher.run(Shared_CQ,conn=Con).get_dataframe()\n",
    "#print('Shared_DF size: ',Shared_DF.shape)\n",
    "\n",
    "# apply filters to exclude cyclic reactions\n",
    "\n",
    "#read molecules to be excluded\n",
    "SheetName=Pathway_Name\n",
    "if SheetName in wb:\n",
    "    mol_df=pd.read_excel(path1+'Exclusion_Molecules.xlsx',sheet_name=Pathway_Name)\n",
    "    Exclude_List=list(map(str,list(mol_df['identifier'])))\n",
    "else:\n",
    "    Exclude_List=[]\n",
    "\n",
    "print('No filters applied: ', Shared_DF.shape)\n",
    "\n",
    "#remove sets in shared entity column\n",
    "df=Shared_DF[~Shared_DF['SharedEntityClass'].isin(['DefinedSet','CandidateSet'])]\n",
    "if Pathway_Name not in ['Metabolism','Digestion and absorption']:\n",
    "    df=Shared_DF[~Shared_DF['SharedEntityClass'].isin(['SimpleEntity'])]\n",
    "print('After removing sets and simple entities: ',df.shape)\n",
    "\n",
    "#select only inter-TopLevelPathway connections\n",
    "df['First_Reaction_Pathway']=df['First_Reaction'].map(PathReacMap_Dict)\n",
    "df['Second_Reaction_Pathway']=df['Second_Reaction'].map(PathReacMap_Dict)\n",
    "df=df[df['First_Reaction_Pathway']!=df['Second_Reaction_Pathway']]\n",
    "#df.drop(['First_Reaction_Pathway','Second_Reaction_Pathway'],axis=1,inplace=True)\n",
    "print('After removing intra-pathway connections: ',df.shape)\n",
    "\n",
    "#remove drug-related events from First_Reaction and Second_Reaction\n",
    "DrugReactions=set(DrugReactions_DF['ReactionID'])\n",
    "df=df[~df['First_Reaction'].isin(DrugReactions)]\n",
    "df=df[~df['Second_Reaction'].isin(DrugReactions)]\n",
    "print('After removing drug events: ',df.shape)\n",
    "\n",
    "#remove exclusion molecules from shared entity column\n",
    "Exclude_ID=set([IDmap_Dict[i] for i in Exclude_List if i in IDmap_Dict])\n",
    "df=df[~df['SharedEntityID'].isin(Exclude_ID)]\n",
    "print('After removing exclusion molecules: ',df.shape)\n",
    "\n",
    "#remove cyclic reactions\n",
    "Index_List=[]\n",
    "com1=df['First_Reaction']+'---'+df['Second_Reaction'] #creating reaction pair\n",
    "com2=df['Second_Reaction']+'---'+df['First_Reaction'] #creating reverse reaction pair\n",
    "#check if reaction pair equals reverse reaction pair\n",
    "for ind,val in enumerate(com1):\n",
    "    if val in list(com2):\n",
    "        Index_List.append(ind)\n",
    "df=df.drop(df.index[Index_List])\n",
    "print('After removing cyclic events: ',df.shape)\n",
    "\n",
    "#find missing reaction connections \n",
    "df['Curated?']=df[['First_Reaction','Second_Reaction']].apply(tuple,1).isin(Preced_DF.apply(tuple,1))\n",
    "print('After removing curated events: ',df[~df['Curated?']].shape)\n",
    "\n",
    "# generate hyperlinks\n",
    "df['First Reaction']=df.First_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(ReacMap_Dict[s])+'\")')\n",
    "df['Second Reaction']=df.Second_Reaction.apply(lambda s: '=HYPERLINK(\"https://reactome.org/content/detail/'+str(s)+'\",\"'+str(ReacMap_Dict[s])+'\")')\n",
    "df['Second Reaction Author']=df.Second_Reaction.apply(lambda a: EventMap_Dict[a])\n",
    "\n",
    "#run query to generate nodes and edges for logic network model\n",
    "Logic_Nodes_CQ=\"\"\"\n",
    "MATCH(tpa:TopLevelPathway)-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "WITH ev\n",
    "OPTIONAL MATCH(ev)-[:input]->(pin:PhysicalEntity)\n",
    "OPTIONAL MATCH (pin)-[:hasMember|hasCandidate*]->(pins:PhysicalEntity)\n",
    "WITH ev,pin,pins\n",
    "OPTIONAL MATCH(ev)-[:output]->(pou:PhysicalEntity)\n",
    "OPTIONAL MATCH (pou)-[:hasMember|hasCandidate*]->(pous:PhysicalEntity)\n",
    "WITH ev,pin,pins,pou,pous\n",
    "OPTIONAL MATCH(ev)-[:catalystActivity|physicalEntity*]->(pca:PhysicalEntity)\n",
    "OPTIONAL MATCH(pca)-[:hasMember|hasCandidate*]->(pcas:PhysicalEntity)\n",
    "WITH ev,pin,pins,pou,pous,pca,pcas\n",
    "OPTIONAL MATCH(ev)-[:regulatedBy|physicalEntity*]->(reg:Regulation)-[:regulator]->(pre:PhysicalEntity)\n",
    "OPTIONAL MATCH(pre)-[:hasMember|hasCandidate*]->(pres:PhysicalEntity)\n",
    "WITH ev, reg, \n",
    "COLLECT(pin.stId)+COLLECT(pins.stId) AS i, COLLECT(pou.stId)+COLLECT(pous.stId) AS o, COLLECT(pca.stId)+COLLECT(pcas.stId) AS c, COLLECT(pre.stId)+COLLECT(pres.stId) AS r\n",
    "RETURN DISTINCT ev.stId AS Reaction, i AS Input, o AS Output, c as Catalyst, r AS Regulator, reg.displayName AS Regulation\n",
    "\"\"\"\n",
    "Logic_Nodes_DF=cypher.run(Logic_Nodes_CQ,conn=Con).get_dataframe()\n",
    "\n",
    "#data processing\n",
    "##creating logic rules and induction/supression value for edges\n",
    "Output=[]\n",
    "Input_List=[y for x in list(Logic_Nodes_DF['Input']) for y in x]\n",
    "InputCounts={i:Input_List.count(i) for i in set(Input_List)}\n",
    "for label,row in df.iterrows():\n",
    "    SharedEntity=row['SharedEntityID']\n",
    "    #label shared entity\n",
    "    if any(SharedEntity in x for x in list(Logic_Nodes_DF['Input'])):\n",
    "        SharedType='Input'\n",
    "    elif any(SharedEntity in x for x in list(Logic_Nodes_DF['Catalyst'])):\n",
    "        SharedType='Catalyst'\n",
    "    elif any(SharedEntity in x for x in list(Logic_Nodes_DF['Regulator'])):\n",
    "        RegulationType=[Logic_Nodes_DF.iloc[i]['Regulation'][:8] for i,v in Logic_Nodes_DF[['Regulator']].iterrows() if SharedEntity in v['Regulator']][0]\n",
    "        SharedType=RegulationType+' Regulator'\n",
    "    else:\n",
    "        print(\"Shared entity not found!\")\n",
    "    #assign Boolean logic and activation/suppression \n",
    "    if SharedType=='Input':\n",
    "        if InputCounts[SharedEntity]==1:\n",
    "            Logic='AND'\n",
    "            Value=1\n",
    "        elif InputCounts[SharedEntity]>1:\n",
    "            Logic='OR'\n",
    "            Value=0\n",
    "    elif SharedType!='Input':\n",
    "        Logic='AND'\n",
    "        if 'Positive' in SharedType:\n",
    "            Value=1\n",
    "        elif 'Negative' in SharedType:\n",
    "            Value=-1\n",
    "        elif 'Catalyst' in SharedType:\n",
    "            Value=1\n",
    "    Output.append([row['First_Reaction_Pathway'],row['First Reaction'],row['Second_Reaction_Pathway'],row['Second Reaction'],row['Second Reaction Author'],SharedEntity,row['SharedEntityClass'],SharedType,Logic,Value,row['Curated?']])\n",
    "Output_df=pd.DataFrame(Output,columns=['First_Reaction_Pathway','First_Reaction','Second_Reaction_Pathway','Second_Reaction','Second_Reaction_Author','Shared_Entity','Shared_Entity_Class','Shared_Entity_Relation_to_Second_Reaction','Logic','Value','Is _It_Curated?'])\n",
    "\n",
    "#write output to file\n",
    "Output_df.to_csv(path2+Pathway_Name+'_IA.csv',index=False,header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Validate algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#validate algorithm\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import gspread\n",
    "from oauth2client.service_account import ServiceAccountCredentials\n",
    "\n",
    "## path\n",
    "notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "path=notebook_path.rsplit('/',1)\n",
    "path1=path[0]+'/Reaction_Connectivity/'\n",
    "path2=path[0]+'/Integrative_Analysis/'\n",
    "\n",
    "# define the scope\n",
    "scope = ['https://spreadsheets.google.com/feeds','https://www.googleapis.com/auth/drive']\n",
    "\n",
    "# add credentials to the account\n",
    "creds = ServiceAccountCredentials.from_json_keyfile_name(path1+'Reactome Connectivity-f994dc05dc97.json', scope)\n",
    "\n",
    "# authorize the clientsheet \n",
    "client = gspread.authorize(creds)\n",
    "\n",
    "# get the instance of the Spreadsheet\n",
    "sheet = client.open('Metabolism_Missing_precedingEvents')\n",
    "\n",
    "# get the first sheet of the Spreadsheet\n",
    "sheet_instance = sheet.get_worksheet(0)\n",
    "\n",
    "# get all the records of the data\n",
    "records_data = sheet_instance.get_all_records()\n",
    "\n",
    "# convert dict to dataframe\n",
    "Reviewed_Data = pd.DataFrame.from_dict(records_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27 rows affected.\n",
      "14371 rows affected.\n",
      "13345 rows affected.\n",
      "20232 rows affected.\n",
      "1264 rows affected.\n",
      "Metabolism\n",
      "2391 rows affected.\n",
      "71229 rows affected.\n",
      "No filters applied:  (71229, 5)\n",
      "After removing sets and simple entities:  (71197, 5)\n",
      "After removing drug events:  (70066, 5)\n",
      "After removing exclusion molecules:  (14621, 5)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import cypher\n",
    "import os\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\" #database connection\n",
    "\n",
    "#run query for pathway name-ID map\n",
    "PathwayName_ID_CQ=\"\"\"\n",
    "MATCH (pa:TopLevelPathway{speciesName:\"Homo sapiens\"})\n",
    "RETURN pa.displayName AS PathwayNames, pa.stId AS PathwayStIds\n",
    "\"\"\" \n",
    "PathwayMap_DF=cypher.run(PathwayName_ID_CQ,conn=Con).get_dataframe()\n",
    "PathwayMap_Dict=PathwayMap_DF.set_index('PathwayNames')['PathwayStIds'].to_dict()\n",
    "\n",
    "#run query for event-author map\n",
    "Event_Author_CQ=\"\"\"\n",
    "MATCH(pa:TopLevelPathway)-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "OPTIONAL MATCH (ev)-[:authored]-(ie:InstanceEdit)-[:author]-(au:Person)\n",
    "RETURN DISTINCT ev.stId AS EventID, au.displayName AS Author\n",
    "\"\"\"\n",
    "EventMap_DF=cypher.run(Event_Author_CQ,conn=Con).get_dataframe()\n",
    "EventMap_Dict=EventMap_DF.set_index('EventID')['Author'].to_dict()\n",
    "\n",
    "#run query for reaction ID-name map\n",
    "ReactionIDName_CQ=\"\"\"\n",
    "MATCH (re:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "RETURN DISTINCT re.stId AS ReactionID, re.displayName AS ReactionName\n",
    "\"\"\"\n",
    "ReacMap_DF=cypher.run(ReactionIDName_CQ,conn=Con).get_dataframe()\n",
    "ReacMap_Dict=ReacMap_DF.set_index('ReactionID')['ReactionName'].to_dict()\n",
    "\n",
    "#run query for reference ID:Reactome ID map\n",
    "IDmap_CQ=\"\"\"\n",
    "MATCH(pa:TopLevelPathway{speciesName:\"Homo sapiens\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|output|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]->(po:PhysicalEntity)-[:referenceEntity]->(ref:ReferenceMolecule)\n",
    "RETURN po.displayName AS Name,ref.identifier AS Ref_Id, po.stId AS React_Id\n",
    "\"\"\"\n",
    "IDmap_DF=cypher.run(IDmap_CQ,conn=Con).get_dataframe()\n",
    "IDmap_Dict=IDmap_DF.set_index('Ref_Id')['React_Id'].to_dict()\n",
    "\n",
    "#run query to get all drug-related reactions\n",
    "DrugReactions_CQ=\"\"\"\n",
    "MATCH(p:TopLevelPathway)-[:hasEvent*]->(r:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:input|output|catalystActivity|hasComponent|hasMember|hasCandidate|physicalEntity|regulatedBy|regulator*]->(d:PhysicalEntity)\n",
    "WHERE (d.schemaClass='ChemicalDrug' OR d.schemaClass='ProteinDrug')\n",
    "RETURN DISTINCT p.displayName AS PathwayName, r.stId AS ReactionID, r.displayName AS ReactionName, d.displayName AS DrugName, d.schemaClass AS DrugType\n",
    "\"\"\"\n",
    "DrugReactions_DF=cypher.run(DrugReactions_CQ,conn=Con).get_dataframe()\n",
    "\n",
    "\n",
    "## path\n",
    "notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "path=notebook_path.rsplit('/',1)\n",
    "path1=path[0]+'/Reaction_Connectivity/'\n",
    "path2=path[0]+'/Integrative_Analysis/'\n",
    "## read exclusion list\n",
    "wb=load_workbook(path1+'Exclusion_Molecules.xlsx',read_only=True)\n",
    "\n",
    "\n",
    "#Analyze missing connections for each top-level pathway\n",
    "m=0\n",
    "#for i in sorted(PathwayMap_Dict):\n",
    "for i in ['Metabolism']:\n",
    "\n",
    "#     # skip big pathways to avoid memory issues\n",
    "#     if i in ['Immune System','Signal Transduction']:\n",
    "#         continue\n",
    "    \n",
    "    print(i)\n",
    "    \n",
    "    #assign pathway ID\n",
    "    Pathway_Name=i\n",
    "    Pathway_stId='\"'+PathwayMap_Dict[i]+'\"'\n",
    "    \n",
    "    #run query to get precedingEvents connections\n",
    "    Preced_CQ=\"\"\"\n",
    "    MATCH(pa:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    MATCH(ev)-[:precedingEvent]->(pe:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    RETURN DISTINCT pe.stId AS First_Reaction, ev.stId AS Second_Reaction\n",
    "    \"\"\"\n",
    "    Preced_DF=cypher.run(Preced_CQ,conn=Con).get_dataframe()\n",
    "    \n",
    "    #run query to get reactions connected by shared entities\n",
    "    Shared_CQ=\"\"\"\n",
    "    ///query for non-set reactions\n",
    "    MATCH(pa1:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po1:PhysicalEntity)\n",
    "    WHERE NOT (po1.schemaClass=\"DefinedSet\" OR po1.schemaClass=\"CandidateSet\" OR po1.stId=\"R-HSA-113595\") //ignore Ub\n",
    "    WITH pa1, ro1, po1\n",
    "    MATCH(pa1)-[:hasEvent*]->(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    WITH ro1, po1, ri1\n",
    "    MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1)\n",
    "    WITH ro1, ri1, po1\n",
    "    RETURN DISTINCT ro1.stId AS First_Reaction, ri1.stId AS Second_Reaction, po1.schemaClass AS SharedEntityClass, po1.displayName AS SharedEntityName, po1.stId AS SharedEntityID\n",
    "    ORDER BY ro1.stId\n",
    "    //query for set connectors\n",
    "    UNION MATCH(pa2:Pathway{stId:\"\"\"+Pathway_stId+\"\"\"})-[:hasEvent*]->(ro2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})-[:output]->(po2a:PhysicalEntity)-[:hasMember|hasCandidate|physicalEntity*]->(po2b:PhysicalEntity)\n",
    "    WHERE (po2a.schemaClass=\"DefinedSet\" OR po2a.schemaClass=\"CandidateSet\") AND NOT (po2a.stId=\"R-HSA-113595\") //ignore Ub\n",
    "    WITH pa2, ro2, po2b\n",
    "    MATCH(pa2)-[:hasEvent*]->(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})\n",
    "    MATCH(po2b)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2)\n",
    "    WITH ro2, ri2, po2b\n",
    "    RETURN DISTINCT ro2.stId AS First_Reaction, ri2.stId AS Second_Reaction, po2b.schemaClass AS SharedEntityClass, po2b.displayName AS SharedEntityName, po2b.stId AS SharedEntityID\n",
    "    ORDER BY ro2.stId\n",
    "    \"\"\"   \n",
    "    Shared_DF=cypher.run(Shared_CQ,conn=Con).get_dataframe()\n",
    "    #print('Shared_DF size: ',Shared_DF.shape)\n",
    "    \n",
    "# apply filters to exclude cyclic reactions\n",
    "\n",
    "    #read molecules to be excluded\n",
    "    SheetName=Pathway_Name\n",
    "    if SheetName in wb:\n",
    "        mol_df=pd.read_excel(path1+'Exclusion_Molecules.xlsx',sheet_name=Pathway_Name)\n",
    "        Exclude_List=list(map(str,list(mol_df['identifier'])))\n",
    "    else:\n",
    "        Exclude_List=[]\n",
    "        \n",
    "    print('No filters applied: ', Shared_DF.shape)\n",
    "    \n",
    "    #remove sets & simple entity in shared entity column\n",
    "    df=Shared_DF[~Shared_DF['SharedEntityClass'].isin(['DefinedSet','CandidateSet'])]\n",
    "    if Pathway_Name not in ['Metabolism','Digestion and absorption']:\n",
    "        df=Shared_DF[~Shared_DF['SharedEntityClass'].isin(['SimpleEntity'])]\n",
    "    print('After removing sets and simple entities: ',df.shape)\n",
    "    \n",
    "    #remove drug-related events from First_Reaction and Second_Reaction\n",
    "    DrugReactions=set(DrugReactions_DF['ReactionID'])\n",
    "    df=df[~df['First_Reaction'].isin(DrugReactions)]\n",
    "    df=df[~df['Second_Reaction'].isin(DrugReactions)]\n",
    "    print('After removing drug events: ',df.shape)\n",
    "    \n",
    "    #remove exclusion molecules from shared entity column\n",
    "    Exclude_ID=set([IDmap_Dict[i] for i in Exclude_List if i in IDmap_Dict])\n",
    "    df=df[~df['SharedEntityID'].isin(Exclude_ID)]\n",
    "    print('After removing exclusion molecules: ',df.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cyclic events:  (879, 5)\n"
     ]
    }
   ],
   "source": [
    "#find cyclic events\n",
    "Index_List=[]\n",
    "com1=df['First_Reaction']+'---'+df['Second_Reaction'] #creating reaction pair\n",
    "com2=df['Second_Reaction']+'---'+df['First_Reaction'] #creating reverse reaction pair\n",
    "#check if reaction pair equals reverse reaction pair\n",
    "for ind,val in enumerate(com1):\n",
    "    if val in list(com2):\n",
    "        Index_List.append(ind)\n",
    "Cyclic_Events=df[df.index.isin(Index_List)]\n",
    "print('Number of cyclic events: ',Cyclic_Events.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-14-8d07956faf5c>:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Cyclic_Events['Combined_Name']=Cyclic_Events['First_Reaction']+'---'+Cyclic_Events['Second_Reaction']\n",
      "<ipython-input-14-8d07956faf5c>:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Checked_Data['Combined_Name']=Checked_Data['First Reaction ID']+'---'+Checked_Data['Second Reaction ID']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Checked_Data=Reviewed_Data[Reviewed_Data['Reason for rejection']=='connects reverse reactions']\n",
    "All_df=df.drop(df.index[Index_List])\n",
    "Cyclic_Events['Combined_Name']=Cyclic_Events['First_Reaction']+'---'+Cyclic_Events['Second_Reaction']\n",
    "All_df['Combined_Name']=All_df['First_Reaction']+'---'+All_df['Second_Reaction']\n",
    "Checked_Data['Combined_Name']=Checked_Data['First Reaction ID']+'---'+Checked_Data['Second Reaction ID']\n",
    "[print(i,': Something wrong!') for i in list(Checked_Data['Combined_Name']) if i in list(All_df['Combined_Name'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Miscellaneous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test1 = [['one', 'two'],\n",
    "['three', 'four', 'five']]\n",
    "testDict = {'one':1,'two':2,'three':3,'four':4,'five':5}\n",
    "print([[testDict[m] for m in n] for n in test1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#example network diagram\n",
    "T = nx.MultiDiGraph()\n",
    "T.add_nodes_from(['one','two','three'])\n",
    "T.add_edges_from([('one','three'),('two','three')])\n",
    "print(nx.info(T))\n",
    "nx.draw_kamada_kawai(T,with_labels=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#example network path\n",
    "T = nx.MultiDiGraph()\n",
    "T.add_nodes_from(['one','two','three'])\n",
    "T.add_edges_from([('three','one'),('two','three'),('two','one')])\n",
    "print(nx.info(T))\n",
    "nx.draw_kamada_kawai(T,with_labels=True)\n",
    "for path in nx.all_simple_paths(T,source='two',target='one'):\n",
    "    print(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#example for creating the adjacency matrix\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "\n",
    "def adjacencyMatrix(nodes,edges):\n",
    "    #create the matrix values with edges\n",
    "    inputNodes = nodes\n",
    "    outputNodes = nodes\n",
    "    edgeMap = []\n",
    "    for inNode in inputNodes:\n",
    "        row = []\n",
    "        for outNode in outputNodes:\n",
    "            if [inNode,outNode] in edges:\n",
    "                row.append(1)\n",
    "            else:\n",
    "                row.append(0)\n",
    "        edgeMap.append(row)\n",
    "    #plot the heatmap\n",
    "    fig, ax = plt.subplots()\n",
    "    im = ax.imshow(edgeMap) #imshow is the heatmap plot\n",
    "#     reactionIds = nodes #vertical axis\n",
    "#     followingRxnsIds = nodes #horizontal axis\n",
    "#     #set ticks\n",
    "#     ax.set_xticks(np.arange(len(reactionIds)))\n",
    "#     ax.set_yticks(np.arange(len(followingRxnsIds)))\n",
    "#     #assign labels\n",
    "#     ax.set_xticklabels(reactionIds)\n",
    "#     ax.set_yticklabels(followingRxnsIds)\n",
    "#     plt.setp(ax.get_xticklabels(),rotation=90,ha='right',rotation_mode='anchor') #allign labels\n",
    "    ax.set_title('Adjacency Matrix of Graph')\n",
    "    ax.set_xlabel('Following Reactions')\n",
    "    ax.set_ylabel('Reactions')\n",
    "    fig.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "adjacencyMatrix(nodeIDs,edgesPreced)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from py2neo import Graph\n",
    "import csv\n",
    "db = Graph(password='neo4jj')\n",
    "\n",
    "pathwayStId='R-HSA-9612973'\n",
    "\n",
    "#query to list reaction based on preceding event relation\n",
    "qPreced = db.run(\n",
    "'MATCH(pa:TopLevelPathway{stId:\"R-HSA-9612973\"})-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})'\n",
    "'OPTIONAL MATCH(ev)-[:precedingEvent]->(pe:ReactionLikeEvent{speciesName:\"Homo sapiens\"})'\n",
    "'RETURN DISTINCT ev.stId AS Event,\"Preceding Event Relation\" AS RelationType,pe.stId AS PrecedingEventCurated'\n",
    ")\n",
    "\n",
    "#query to list reactions where the output of one reaction is checked for being the input, catalyst activity or regulator of another\n",
    "#if the output is a set, the members are split\n",
    "#if output is SimpleEntity, it is ignored\n",
    "qShared = db.run(\n",
    "'MATCH(pa:TopLevelPathway{stId:\"R-HSA-9612973\"})-[:hasEvent*]->(ro:ReactionLikeEvent{speciesName:\"Homo sapiens\"})'\n",
    "'WITH pa, ro'\n",
    "'MATCH (ro)-[:output]->(po1:PhysicalEntity)'\n",
    "'//WHERE NOT (po1.schemaClass=\"SimpleEntity\"  OR po1.stId=\"R-HSA-184258\" OR po1.stId=\"R-HSA-3149456\" OR po1.stId=\"R-HSA-917706\")'\n",
    "'WHERE NOT (po1.schemaClass=\"SimpleEntity\")'\n",
    "'WITH pa, ro, po1'\n",
    "'OPTIONAL MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity*]-(ri1:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)'\n",
    "'OPTIONAL MATCH(po1)<-[:input|catalystActivity|regulatedBy|regulator|physicalEntity|hasMember|hasCandidate*]-(ri2:ReactionLikeEvent{speciesName:\"Homo sapiens\"})<-[:hasEvent*]-(pa)'\n",
    "'WITH pa, ro, po1, ri1, ri2'\n",
    "'RETURN DISTINCT pa.displayName AS PathwayName,ro.stId AS Event,\"Shared entity relation\" AS RelationType,\\\n",
    "CASE\\\n",
    "WHEN po1.schemaClass=\"DefinedSet\"\\\n",
    "THEN ri2.stId\\\n",
    "ELSE ri1.stId\\\n",
    "END AS InputReaction'\n",
    ")\n",
    "\n",
    "reactList = []\n",
    "for react in qPreced:\n",
    "    lreact=list(react)\n",
    "    orderChan=[2,1,0]\n",
    "    lreact=[lreact[i] for i in orderChan]\n",
    "    reactList.append(lreact)\n",
    "reactList.insert(0,[\"First Reaction\",\"Type\",\"Second Reaction\"])\n",
    "\n",
    "path = 'C:\\\\Users\\\\tvarusai\\\\EBI_Work\\\\Reactome\\\\Network Analysis Project\\\\ReactomeNetworkAnalysis\\\\'\n",
    "fileName = 'graphPreced.csv'\n",
    "with open(path+fileName,'w',newline='') as file:\n",
    "    fileWriter=csv.writer(file)\n",
    "    fileWriter.writerows(reactList)\n",
    "    \n",
    "reactList = []\n",
    "for react in qShared:\n",
    "    lreact=list(react)\n",
    "    reactList.append(lreact)\n",
    "reactList.insert(0,[\"First Reaction\",\"Type\",\"Second Reaction\"])\n",
    "\n",
    "path = 'C:\\\\Users\\\\tvarusai\\\\EBI_Work\\\\Reactome\\\\Network Analysis Project\\\\ReactomeNetworkAnalysis\\\\'\n",
    "fileName = 'graphShared.csv'\n",
    "with open(path+fileName,'w',newline='') as file:\n",
    "    fileWriter=csv.writer(file)\n",
    "    fileWriter.writerows(reactList)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from neo4j import GraphDatabase\n",
    "uri = 'bolt://localhost:7687'\n",
    "driver = GraphDatabase.driver(uri,auth=('neo4j','neo4jj'))\n",
    "def precedConnect(tx, stableID):\n",
    "    for record in tx.run('MATCH(pa:TopLevelPathway{stId:{stableID})-[:hasEvent*]->(ev:ReactionLikeEvent{speciesName:\"Homo sapiens\"})'\n",
    "                         'OPTIONAL MATCH(ev)-[:precedingEvent]->(pe:ReactionLikeEvent{speciesName:\"Homo sapiens\"})'\n",
    "                         'RETURN DISTINCT ev.stId AS Event,\"Preceding Event Relation\" AS RelationType,pe.stId AS PrecedingEventCurated',stableID=stableID):\n",
    "        print(record['pe.stId'])\n",
    "\n",
    "with driver.session() as session:\n",
    "    session.read_transaction(precedConnect, \"R-HSA-9612973\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "path = 'C:\\\\Users\\\\tvarusai\\\\EBI_Work\\\\Reactome\\\\Network Analysis Project\\\\ReactomeNetworkAnalysis\\\\'\n",
    "nodeList = 'Autophagy_Reactions_Nodes_Reactions.csv'\n",
    "edgeList = 'Autophagy_Reactions_Edges_Reactions.csv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test run Neo4j cypher queries from Jupyter notebook\n",
    "from py2neo import Graph,Node,Relationship\n",
    "graph=Graph('http://localhost:7474/browser/',auth=('neo4j','reactome'))\n",
    "AllPathways=\"MATCH (pa:TopLevelPathway) RETURN COUNT(pa.displayName)\"\n",
    "AllPathwaysResults=graph.evaluate(AllPathways)\n",
    "AllPathwaysResults"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install ipython-cypher"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext cypher"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cypher http://neo4j:reactome@localhost:7474/db/data MATCH (pa:TopLevelPathway{speciesName:'Homo sapiens'})-[:hasEvent*]->(re:ReactionLikeEvent) RETURN DISTINCT pa.displayName AS PathwayName, COUNT(DISTINCT re.stId) AS ReactionCount"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%cypher http://neo4j:reactome@localhost:7474/db/data\n",
    "MATCH (pa:TopLevelPathway{speciesName:'Homo sapiens'})-[:hasEvent*]->(re:ReactionLikeEvent)\n",
    "RETURN DISTINCT pa.displayName AS PathwayName, COUNT(DISTINCT re.stId) AS ReactionCount"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### WORKING SCRIPT\n",
    "\n",
    "\n",
    "import cypher\n",
    "\n",
    "Con=\"http://neo4j:reactome@localhost:7474/db/data\"\n",
    "cq=\"\"\"\n",
    "MATCH (pa:TopLevelPathway{speciesName:'Homo sapiens'})-[:hasEvent*]->(re:ReactionLikeEvent)\n",
    "RETURN DISTINCT pa.displayName AS PathwayName, COUNT(DISTINCT re.stId) AS ReactionCount\n",
    "\"\"\"\n",
    "cq=\"\"\"\n",
    "MATCH (pa:TopLevelPathway{speciesName:'Homo sapiens'})-[:hasEvent*]->(re:ReactionLikeEvent)\n",
    "RETURN DISTINCT pa.displayName AS PathwayName, COUNT(DISTINCT re.stId) AS ReactionCount\n",
    "\"\"\"\n",
    "\n",
    "results=cypher.run(cq,conn=Con).get_dataframe()\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#working script for logic network model\n",
    "#imports\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "#path\n",
    "notebook_path = os.path.abspath(\"NetworkAnalysis_1.ipynb\")\n",
    "path=notebook_path.rsplit('/',1)\n",
    "path=path[0]+'/Reaction_Logic/'\n",
    "\n",
    "#read files\n",
    "PathwayName='CircadianClock' #pathway name\n",
    "##Nodes\n",
    "NodesFile=PathwayName+'_LogicNetwork_Nodes.csv'\n",
    "Nodes=pd.read_csv(path+NodesFile)\n",
    "##Edges\n",
    "EdgesFile=PathwayName+'_LogicNetwork_Edges.csv'\n",
    "Edges=pd.read_csv(path+EdgesFile)\n",
    "\n",
    "#data processing\n",
    "##creating logic rules and induction/supression value for edges\n",
    "BoolEdgesList=[]\n",
    "OutputCounts=Nodes['Output'].value_counts()\n",
    "for label,row in Edges.iterrows():\n",
    "    SharedEntity=row['SharedEntity']\n",
    "    SharedType=row['SharedType']\n",
    "    if SharedType=='Input':\n",
    "        if OutputCounts[SharedEntity]==1:\n",
    "            Logic='AND'\n",
    "            Value=1\n",
    "        elif OutputCounts[SharedEntity]>1:\n",
    "            Logic='OR'\n",
    "            Value=0\n",
    "    elif SharedType!='Input':\n",
    "        Logic='AND'\n",
    "        if 'Positive' in SharedType:\n",
    "            Value=1\n",
    "        elif 'Negative' in SharedType:\n",
    "            Value=-1\n",
    "        elif 'Catalyst' in SharedType:\n",
    "            Value=1\n",
    "    BoolEdgesList.append([row['First_Reaction'],row['Second_Reaction'],Logic,Value])\n",
    "BoolEdges=pd.DataFrame(BoolEdgesList,columns=['First_Reaction','Second_Reaction','Logic','Value'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
